% template.tex, dated April 5 2013
% This is a template file for Annual Reviews 1 column Journals
%
% Compilation using ar-1col.cls' - version 1.0, Aptara Inc.
% (c) 2013 AR
%
% Steps to compile: latex latex latex
%
% For tracking purposes => this is v1.0 - Apr. 2013

\documentclass{style/ar-1col}
\usepackage[utf8]{inputenc}
\usepackage{url}
\usepackage[numbers]{natbib}
\bibliographystyle{bib/ar-style6}
\usepackage{multirow}
\usepackage{todonotes}
%\usepackage{caption}
\usepackage{subcaption}
\usepackage{graphicx}
\usepackage{comment}
\usepackage{tabularx}
\usepackage{tablefootnote}

% Set up ORCiD. See <https://tex.stackexchange.com/a/445583>
\begin{comment}
\usepackage{scalerel}
\usepackage{tikz}
\usetikzlibrary{svg.path}
\definecolor{orcidlogocol}{HTML}{A6CE39}
\tikzset{
  orcidlogo/.pic={
    \fill[orcidlogocol] svg{M256,128c0,70.7-57.3,128-128,128C57.3,256,0,198.7,0,128C0,57.3,57.3,0,128,0C198.7,0,256,57.3,256,128z};
    \fill[white] svg{M86.3,186.2H70.9V79.1h15.4v48.4V186.2z}
                 svg{M108.9,79.1h41.6c39.6,0,57,28.3,57,53.6c0,27.5-21.5,53.6-56.8,53.6h-41.8V79.1z M124.3,172.4h24.5c34.9,0,42.9-26.5,42.9-39.7c0-21.5-13.7-39.7-43.7-39.7h-23.7V172.4z}
                 svg{M88.7,56.8c0,5.5-4.5,10.1-10.1,10.1c-5.6,0-10.1-4.6-10.1-10.1c0-5.6,4.5-10.1,10.1-10.1C84.2,46.7,88.7,51.3,88.7,56.8z};
  }
}
\newcommand\orcidicon[1]{\href{https://orcid.org/#1}{\mbox{\scalerel*{
\begin{tikzpicture}[yscale=-1,transform shape]
\pic{orcidlogo};
\end{tikzpicture}
}{|}}}}
\end{comment}

\newcommand\orcidicon[1]{\href{https://orcid.org/#1}{$^\sigma$}}

\usepackage{hyperref} %<--- Load after everything else

% Define a way to rotate table headings. See
% <https://tex.stackexchange.com/a/98439>
\newcommand*\rot{\rotatebox{90}}

% Define a way to cram two lines into a non-paragraph cell
% See <https://tex.stackexchange.com/a/132661>
\newcommand{\twoline}[2]{\vtop{\hbox{\strut #1}\hbox{\strut #2}}}

\listfiles

\setcounter{secnumdepth}{4}

% Metadata Information
\jname{Xxxx. Xxx. Xxx. Xxx.}
\jvol{AA}
\jyear{YYYY}
\doi{10.1146/((please add article doi))}


% Document starts
\begin{document}

\newcommand{\shorttitle}{Pangenome graphs}
\newcommand{\firstauthorlastname}{Eizenga}

% Page header
\markboth{\firstauthorlastname\ et al.}{\shorttitle}

% Title
\title{\shorttitle}


%Authors, affiliations address.
\author{%
  Jordan M \firstauthorlastname$^1$\orcidicon{0000-0001-8345-8356},
  Adam M Novak$^1$\orcidicon{0000-0001-5828-047X},
  \\
  Jonas A Sibbesen$^1$\orcidicon{0000-0002-5528-0236},
  Simon Heumos$^2$\orcidicon{0000-0003-3326-817X},
  \\
  Ali Ghaffaari$^{5,6,7}$,
  Glenn Hickey$^1$,
  Xian Chang$^1$,
  \\
  Josiah D Seaman$^{3,4}$\orcidicon{0000-0003-2374-630X},
  Robin Rounthwaite$^1$\orcidicon{0000-0002-0264-7487},
  Jana Ebler$^{5,6,7}$,
  Mikko Rautiainen$^{5,6,7}$\orcidicon{0000-0003-2971-267X},
  \\
  Shilpa Garg$^{8,9,10}$,
  Benedict Paten$^1$\orcidicon{0000-0001-8863-3539},
  \\
  Tobias Marschall$^{5,6}$\orcidicon{0000-0002-9376-1030},
  Jouni Sir\'{e}n$^1$\orcidicon{0000-0001-5828-4139},
  \\
  and Erik Garrison$^{*1}$\orcidicon{0000-0003-3821-631X}
\affil{\tiny$^1$Genomics Institute, University of California Santa Cruz, Santa Cruz, CA, USA, 95064}
\affil{\tiny$^2$Quantitative Biology Center, University of Tübingen, Tübingen, Germany, 72076}
\affil{\tiny$^3$Royal Botanic Gardens Kew, Richmond, UK, TW9 3AB}
\affil{\tiny$^4$Queen Mary University of London, London, UK, E1 4NS}
\affil{\tiny$^5$Center for Bioinformatics, Saarland University, Saarbr\"{u}cken, Germany, 66123}
\affil{\tiny$^6$Max Planck Institute for Informatics, Saarbr\"{u}cken, Germany, 66123}
\affil{\tiny$^7$Saarbr\"{u}cken Graduate School for Computer Science, Saarbr\"{u}cken, Germany, 66123}
\affil{\tiny$^8$Department of Genetics, Harvard Medical School, Boston, MA 02215}
\affil{\tiny$^9$Department of Data Sciences, Dana-Farber Cancer Institute, Boston, MA 02215}
\affil{\tiny$^{10}$Department of Biomedical Informatics, Harvard Medical School, Boston, MA 02215}
\affil{\tiny$\sigma$ORCID identifier hyperlinks}
\affil{\tiny$^*$email: erik.garrison@ucsc.edu}
}

%Abstract
\begin{abstract}
  Improved sequencing technology lets us collect whole genome assemblies of high quality from many individuals, which opens the possibility of working with a pangenome that represents the total genomic information in a species.
  Many recently-developed methods use pangenome reference systems as precise bases for bioinformatic analyses.
  These methods use compressed, often graph-based, and sometimes approximate models to implement algorithms for sequence alignment, visualization, functional genomics, and association studies at the scale of large eukaryotic pangenomes.
  Although sequence graphs feature prominently in these methods, most methods focus on communicating results in the frame of a linear reference genome, suggesting that linear reference representations will remain relevant into the near future.
  However, graphical pangenomes can work harmoniously with multiple linear reference systems, and so we expect them to unify and build upon standard reference genomes, rather than replacing them.
\end{abstract}

%Keywords, etc.
\begin{keywords}
genome graph, variation graph, graph, pangenome, reference
\end{keywords}
\maketitle

%Table of Contents
%\tableofcontents

\input{sections/intro}
\input{sections/models}
\input{sections/relating}
\input{sections/formats}
\input{sections/applications}
\input{sections/discussion}
\input{sections/endmatter}
\input{sections/figures}
\input{sections/tables}

% References
%
% Margin notes within bibliography

\noindent
\bibliography{bib/references.bib}

%To download the appropriate bibliography style file, please see \url{http://www.annualreviews.org/page/authors/author-instructions/preparing/latex}. 


%Please see the Style Guide document for instructions on preparing your Literature Cited.

%The citations should be numbered in alphabetical order, with titles. For example:



\end{document}
