\section{Applications of pangenomic models}

\subsection{Error correction}
% Robin

\subsection{Variant calling and genotyping}
% Glenn

Genome graphs can leverage known sequence variation, available in increasingly large public databases, to improve variant identification in new samples.
In general, variants are determined by finding paths through the graph that are best supported by reads.
Projecting the paths back to the reference yields the variants, typically in VCF format.
When only paths from graph are considered, this process is considered \emph{genotyping}.
Novel variants can be \emph{called} by augmenting the paths with edits from the reads.

Graphtyper \cite{eggertsson2017graphtyper} begins with a pan-genome graph created from variants in dbSNP, and iteratively updates it with variants discovered in input reads.
Its authors showed it to be more accurate on benchmarks than top linear reference-based approaches such as GATK.
It was applied to WGS data from 28,075 samples from an Icelandic population study.
Another approach used a graph derived from the 1000 Genomes Project \cite{Rakocevic_2019} to improve calling accuracy, as well as find variants absent from common benchmarks such as GIAB.

Difficult and diverse genomic regions stand to gain the most by moving from avlinear reference to genome graphs.
To this end, much work has been done exploring the application of graphs to HLA typing.
MHC/HLA bla bla bla.
The Population Reference Graph (PRG) \cite{dilthey2015improved} is created from a multiple sequence alignment (MSA) of know haplotypes and variants spanning the MHC region.
Reads are mapped to to the graph, then used to extract a pair of ``chromotype'' paths that best represent them.
The chromotypes are then used as linear references for standard mapping and calling tools, and the results projected back to the linear reference.
PRG*HLA \cite{dilthey2018hla} explicitly models the HLA genes, rather than the whole MHC region.
HLA types are imputed directly from the graph, rather than intermediate chromotypes, providing faster performance and  HLA*LA \cite{dilthey2019hla} further improves runtime and adds support for long read and assembly inputs.
Kourami \cite{lee2018kourami} is another method for HLA typing that assembles haplotypes using a graph genome as a guide, and is able to incorporate novel variation from the reads. 

ExpansionHunter \cite{dolzhenko2019expansionhunter} uses genome graphs to model low complexity regions in the genome.
To this end, regular expressions are used to define variable sites of tandem repeats, which come out as self-loops in the graph.
The authors show that short read data can be used to more accurately type clinically relevant sites of such varaition using this approach than standard variant calling pipelines.

Structural variants (SVs) are mutational events of at least 50bp in length.
SVs can be increasingly well characterized by long read sequencing, but this technology remains prohibitively expensive for population scale studies.
Known SVs can be naturally represented as genome graphs, and genotyped using WGS short read data.
Paragraph \cite{chen2019paragraph} is a SV genotyper that operates on a genome graphs constructed from each variant in a VCF of SV calls.
For each variant, it remaps nearby reads as retrieved from a linear alignment, to the graph.
It then computes a genotype from the support of each allele's breakpoints in the graph alignment.
Paragraph was shown to outperform similar methods that rely on linear references by a wide margin.
vg's SV \cite{hickey2019genotyping} genotyper can be run on any genome graph, provided it contains embedded paths that can be used for coordinates.
Unlike all other methods mentioned above, which remap a subset of relevant reads to the graph, vg operates directly on a graph alignment of the entire read set.
It uses the snarl decomposition \cite{paten2018superbubbles} to identify sites of variation in the graph, and derives haplotypes using read support.
It was shown to be much more accurate than linear reference-based approaches, and was also evaluated on graphs derived from alignments of assemblies.


oops need to add Bayestyper \cite{sibbesen2018accurate}


\subsection{Assembly and metagenomics}
% Erik

\subsection{Epigenomics}
% Glenn

These methods use graph genomes to reduce reference bias and improve peak calling \cite{grytten2019graph}  \cite{groza2019personalized}.  

\subsection{Transcriptomics}

Using genome graphs for the analyses of transcriptomic sequencing data have generally received less attention compared to its whole genome sequencing counterpart.
However, for estimating allele-specific expression (ASE), reference-bias is an important issue that needs to be considered in order to get reliable results \cite{Degner2009-vw,Castel2015-ef}.
Here the expression levels of genes or transcripts are estimated on each allele separately by counting the difference in number of mapped RNA sequencing (RNA-seq) reads across heterozygous variants.
A bias towards one of the alleles can therefore result in an false-positive difference in expression between the alleles.

Filtering strategies based on simulations that remove variant sites or reads with likely bias have been proposed as a way to mitigate this, but these approaches can result in less detectable allele-specific expressed genes or lower expression values \cite{Castel2015-ef,Van_de_Geijn2015-dz}.
Variant-aware methods, such as genome graphs, provides a means that does not rely on filtering and instead takes the sequence difference between the two alleles into account during mapping.
The simplest of these involves creating a personalized diploid genome or transcriptome and using standard linear mapping method.
This approach has shown to reduce reference-bias and improve estimate of ASE \cite{Rozowsky_2011,Raghupathy2018-sd}.
One major limitation of these methods is that they require that phased haplotypes are available for the samples of interest 
ASElux ... \cite{Miao2018-ps}
Variant-aware mapping methods on the other hand allow for using unphased variation and known variants from previous studies.
GSNAP ... \cite{Wu2010-hv}
iMapSplice ... \cite{Liu_2018}
HISAT2 was the first genome graph based method capable of splice-aware mapping of RNA-seq data \cite{Kim_2019}. 
Using a combination of a hierarchical graph FM index and a repeat region index (see ? section for more details) HISAT2 is able to map RNA-seq reads at an impressive speed to a genome graph containing 14.5M variations without sacrificing accuracy. 
Furthermore, the method is able to use known splice-junctions and discover novel junctions using splice-site information.
Their benchmark, however, only shows results for DNA sequencing data and therefore it is unknown how well this variant-aware version reduces reference-bias for RNA-seq data.
Recently, the ability to create spliced variation graphs was added to the vg toolkit \cite{Garrison_2018}. 
In these graphs splice-junctions are added as edges to a variation graph similar to the addition of a deletion event. 
This allows for the vg toolkit to also map RNA-seq data, however the approach is limited to only known junctions.

Similar to WGS, variation-aware analysis of RNA-seq data is important for estimating expression of genes and transcripts in highly polymorphic regions, such as HLA genes in the human major histocompatibility complex region. 
It has been shown that estimation of expression can be improved by comparing the reads against a collection of known HLA haplotypes instead of the linear reference \cite{Lee2018-mm,Aguiar2019-fy}.
AltHapAlignR is one such method \cite{Lee2018-mm}. It first maps the RNA-seq reads to the linear reference. Unmapped reads and reads that map to the HLA region are then collected and mapped to 7 alternative haplotypes followed by expression estimation in a haplotype-specific manner. 
While this approach improves estimation of ASE it is unclear whether the method can scale to the thousands of haplotypes known in the HLA region.
HLApers on the other hand first alings the RNA-seq reads against an index containing thousands of known HLA haplotypes using either a regular mapper or pseudoaligner \cite{Aguiar2019-fy}.
These aligned are then used predicts the most likely haplotype pair for each gene to which the reads are aligned against a second time and from which expression is estimated.  
Using this approach they were able to align more reads, especially for genes with high sequence divergence from the reference. 
One limitation of this approach is that it only aligns to known haplotypes in the HLA region and not the whole genome as AltHapAlignR thereby not accounting for HLA pseudogenes.
Both of these methods does not use a genome graph representation, however, the collection of haplotypes could easily be represented using a graph.
Known haplotypes could then be encoded using the GBWT index allowing it to scale to whole genome and potentially hundreds of thousands of haplotypes \cite{Siren_2019}. 

Using variation in the intronic regions can even be beneficial when analysing RNA-seq data. 
Variation in these regions can either disrupt or create new splice-site motifs resulting in intron retention or novel splicing, respectively. 
This can affect mapping algorithms, since the absence or presence of canonical splice-site motifs are often used to score alignments across splice-junctions. 
Indeed, by using a personalized genome approach Stein \textit{et al.} was able to identify 506 personal splice-junctions in 75 individuals of which 437 was novel \cite{Stein_2015}. 
Later, Lui \textit{et al.} more than tripled this number in the same individuals by detecting many more novel canonical personal splice-junctions using iMapSplice \cite{Liu_2018}.
Both of these methods thereby shows the importance of variant-aware analysis for reference-bias free splice-junction detection. 


% Papers that maybe needs to be added:
% Beretta \textit{et al.}
% Denti \textit{et al.}
% Bray \textit{et al.}

