\section{Applications of pangenomic models}

\subsection{Error correction}
% Robin

\subsection{Variant calling and genotyping}
% Glenn

Genome graphs can leverage known sequence variation, available in increasingly large public databases, to improve variant identification in new samples.
In general, variants are determined by finding paths through the graph that are best supported by mapped reads.
Projecting these paths back to the reference yields the variants, typically in VCF format.
When only paths from graph are considered, this process is considered \it{genotyping}.
Novel variants can be \it{called} by augmenting the paths with edits from the reads.

GraphTyper \cite{eggertsson2017graphtyper} begins with a pan-genome graph created from variants in dbSNP, and iteratively updates it with variants discovered in input reads.
Its authors showed it to be more accurate on benchmarks than top linear reference-based approaches such as GATK.
It was applied to WGS data from 28,075 samples from an Icelandic population study.
Another approach used a graph derived from the 1000 Genomes Project \cite{Rakocevic_2019} to improve calling accuracy, as well as find variants absent from common benchmarks such as GIAB.

Difficult and diverse genomic regions stand to gain the most by moving from a linear reference to genome graph.
To this end, much work has been done exploring the application of graphs to the major histocompatibility complex (MHC) region on chromosome 6, which contains the human leukocyte antigen (HLA) genes.
Determining the HLA alleles in a sample, or HLA typing, is often of great importance, but due to high polymorphism within the MHC region, WGS mapped to the linear reference cannot be effectively used and more costly Sanger sequencing is required.
The Population Reference Graph (PRG) \cite{dilthey2015improved} is created from a multiple sequence alignment (MSA) of know haplotypes and variants spanning the MHC region.
Reads are mapped to to the graph then used to extract a pair of ``chromotype'' paths that best represent them, which are then used linear references for standard mapping and calling tools.
PRG*HLA \cite{dilthey2018hla} explicitly models the HLA genes, rather than the whole MHC region.
HLA types are imputed directly from the graph, rather than intermediate chromotypes, providing faster performance and  HLA*LA \cite{dilthey2019hla} further improves runtime and adds support for long read and assembly inputs.
Kourami \cite{lee2018kourami} is another method for HLA typing that assembles haplotypes using a graph genome as a guide, and is able to incorporate novel variation from the reads.
HLA typing with these methods using WGS has been shown to be as accurate as the conventional Sanger sequencing based approach.

ExpansionHunter \cite{dolzhenko2019expansionhunter} uses genome graphs to model low complexity regions in the genome.
It uses regular expressions are used to define variable sites of tandem repeats, which are modeled as self-loops in the graph.
The authors show that short read data can be used to more accurately type clinically relevant sites of such variation using this approach than standard variant calling pipelines.

Structural variants (SVs) are mutational events of at least 50bp in length.
SVs can be increasingly well characterized by long read sequencing, but this technology remains prohibitively expensive for population scale studies.
Known SVs can be naturally represented as genome graphs, and genotyped using WGS short read data.
Bayestyper \cite{sibbesen2018accurate} uses the distribution of kmers from sequencing reads, as well as that of the graph, to accurately identify variant haplotypes.  
Paragraph \cite{chen2019paragraph} is a SV genotyper that operates on a genome graphs constructed from each variant in a VCF of SV calls.
For each variant, it remaps nearby reads as retrieved from a linear alignment, to the graph.
It then computes a genotype from the support of each allele's breakpoints in the graph alignment.
Paragraph was shown to outperform similar methods that rely on linear references by a wide margin.
vg's SV \cite{hickey2019genotyping} genotyper can be run on any genome graph, provided it contains embedded paths that can be used for coordinates.
Unlike all other methods mentioned above, which remap a subset of relevant reads to the graph, vg operates directly on a graph alignment of the entire read set.
It uses the snarl decomposition \cite{paten2018superbubbles} to identify sites of variation in the graph, and derives haplotypes using read support.
It was shown to be much more accurate than linear reference-based approaches, and was also evaluated on graphs derived from alignments of assemblies.

\subsection{Assembly and metagenomics}
% Erik

\subsection{Epigenomics}
% Glenn

CHiP-seq data, reads that bind to specific transcription factors, are mapped back to the reference genome in order to locate binding sites.  
Graph Peak Caller is based on vg and is the first tool to use a genome graph for this process \cite{grytten2019graph}.
It was shown to find binding sites more enriched for known DNA binding motifs than linear methods on \emph{A. thaliana}.
It was also applied to human data to discover novel sites for enhancers in the human genome \cite{groza2019personalized}. 

\subsection{Transcriptomics}

Using genome reference graphs for the analyses of transcriptomic sequencing data have generally received less attention compared to its whole genome sequencing counterpart.
However, reference-bias is an important issue that needs to be addressed when estimating allele-specific expression (ASE \cite{Degner2009-vw,Castel2015-ef}.
Here the expression levels of genes or transcripts are estimated on each allele separately by counting the difference in number of mapped RNA sequencing (RNA-seq) reads across heterozygous variants.
A bias towards one of the alleles can therefore result in an false-positive difference in expression between the alleles. \\

Filtering strategies based on simulations that remove variant sites or reads with likely bias have been proposed as a way to mitigate this, but these approaches can result in less detectable allele-specific expressed genes or lower expression values \cite{Castel2015-ef,Van_de_Geijn2015-dz}. 
Variant-aware methods, such as genome graphs, provides a means that does not rely on filtering and instead takes the sequence difference between the two alleles into account during mapping.
The simplest approach involves creating a personalized diploid genome or transcriptome followed by the use of standard linear mapping method \cite{Turro2011-op,Rozowsky_2011,Bray_2016,Raghupathy2018-sd}.
Methods using this approach have shown to reduce reference-bias and improve estimation of ASE, but one limitation is that they require that phased haplotypes are available for the samples investigated.
Variant-aware mapping methods on the other hand does not require that the variants are phased.
GSNAP was the first variation- and splicing-aware mapping method developed for RNA-seq data \cite{Wu2010-hv}.
It uses a kmer-based approach where both the genome and a set of SNVs are indexed using hash tables. % We might want to add GSNAP introduction to the model section since it also works on WGS.
Current version uses a suffix array in addition to the hash table.
Alignment is performed by searching for kmers in the read that are in close proximity on the reference genome. % Needs a much more detailed and better description. 
GSNAP is still competitive with regards to mapping accuracy, but it is generally much slower than contemporary mapping methods.
The authors does not specifically show that GSNAP reduces reference-bias, however it has been shown later by others \cite{Castel2015-ef}.
ASElux uses all heterozygote exonic SNVs in an individual to create a suffix array index of the alleles and their flanking sequences \cite{Miao2018-ps}. 
This index is used to filter read pairs that does not overlap any SNVs with up to 2 mismatches allowed in the flanking regions. 
Read pairs that pass this filter are then aligned to a different suffix array index consisting of exonic and intronic regions in order to find its exact location. 
ASElux showed better reduction in reference-bias and higher alignment speed compared to variation aware-mapping methods.
Similar to ASElux, iMapSplice creates an index of the SNV alleles and flanking sequences \cite{Liu_2018}.
The sequences are indexed using enhanced suffix arrays and reads are mapped to both this index and the reference genome using semi-maximal prefix matching.
The authors showed that iMapSplice achieves higher mapping accuracy and lower reference-bias compared to both linear and other variant-aware mapping methods.
GSNAP, ASElux and iMapSplice only supports SNVs and is therefore not able to reduce reference-bias around indels.
HISAT2 was the first genome graph based method capable of splice-aware mapping of RNA-seq data that also supports non-SNVs \cite{Kim_2019}. 
Their method can use SNVs, deletions of any length and insertions up to 20 bp.
Using a combination of a hierarchical graph FM index and a repeat region index (see ? section for more details) HISAT2 is able to map RNA-seq reads at an impressive speed to a genome graph containing 14.5M variations without sacrificing accuracy. 
Their benchmark, however, only shows results for DNA sequencing data and therefore it is unknown how well this variant-aware version reduces reference-bias for RNA-seq data.
Recently, the ability to create spliced variation graphs was added to the vg toolkit \cite{Garrison_2018}. 
In these graphs known splice-junctions are added as edges to a variation graph similar to the addition of a deletion event while transcripts are embedded as paths through the graph. 
This allows for the current algorithms in vg to also be used for mapping of RNA-seq data. 
vg supports any type of variation, but it is limited to only known splice-junctions and thus reads that span a novel junction will only map partially. \\

Similar to WGS, variation-aware analysis of RNA-seq data is important for estimating expression of genes and transcripts in highly polymorphic regions, such as HLA genes in the human major histocompatibility complex region. 
It has been shown that estimation of expression can be improved by comparing the reads against a collection of known HLA haplotypes instead of the linear reference. 
AltHapAlignR is one such method \cite{Lee2018-mm}. It first maps the RNA-seq reads to the linear reference. Unmapped reads and reads that map to the HLA region are then collected and mapped to 7 alternative haplotypes followed by expression estimation in a haplotype-specific manner. 
While this approach improves estimation of ASE it is unclear whether the method can scale to the thousands of haplotypes known in the HLA region.
HLApers on the other hand first aligns the RNA-seq reads against an index containing thousands of known HLA haplotypes using either a regular mapper or pseudoaligner \cite{Aguiar2019-fy}.
The aligned reads are then used to predict the most likely haplotype pair for each gene to which the reads are aligned against a second time and from which expression is estimated.  
Using this approach they were able to align more reads compared to using the linear reference, especially for genes with high sequence divergence from the reference, but only aligning to the HLA haplotypes might bias their results toward these. \\

Using variation in the intronic regions can also be beneficial when analysing RNA-seq data. 
Variation in these regions can either disrupt or create new splice-site motifs resulting in intron retention or novel splicing, respectively. 
This can affect mapping algorithms, since the absence or presence of canonical splice-site motifs are often used to score alignments across splice-junctions. 
Indeed, by using a personalized genome approach Stein \textit{et al.} was able to identify 506 personal splice-junctions in 75 individuals of which 437 was novel \cite{Stein_2015}. 
Later, Lui \textit{et al.} more than tripled this number in the same individuals using iMapSplice, thereby showing the importance of variant-aware analysis for bias-free splice-junction detection \cite{Liu_2018}.

% Papers that maybe needs to be added here or under calling/genotyping:
% HLA typing from RNA-seq (Beretta \textit{et al.}, Denti \textit{et al.})


