\section{Applications of pangenomic models}
\label{sec:applications}

Although graphical pangenomic techniques can be applied throughout biology, most recent work has focused on a handful of applications where they can provide substantial benefits.
Reduced reference bias and coherent representations of alleles yields significant improvements in structural variation detection and can decrease the runtime costs of genotyping.
They have enabled haplotype reconstruction of quasispecies in metagenomic studies and transcripts in functional genomics, and supported pangenomic association studies.

\subsection{Variant calling and genotyping}
\label{sec:variantcalling}

Typically, \emph{variant calling} and \emph{genotyping} indicate different aspects of a reference-guided genome inference process.
Genotyping consists of determining whether a previously observed variant is present in a new sample, whereas variant calling involves detecting yet unobserved variation.
When our reference system is a linear genome, these two steps are often merged.
A single process will detect candidate variation and infer a sample genotype at each putatively variable locus \cite{Li_2011,Garrison_2012}.

These methods are often Bayesian, and combine a model for observation and error with a simple \emph{a priori} model of the pattern of variation that we expect to see, typically based on the expected rate of heterozygosity or mutation.
By using \emph{multisample} variant calling, genotype phasing, and genotype imputation \cite{1000_2015,browning2011haplotype}, we can share information between samples to improve the accuracy of our reconstruction of genomes from short reads.
However, this joint calling approach is expensive, and not applicable when we only have a few new genomes to reconstruct.
Furthermore, it does not help our primary interpretation of new sequencing data during read mapping.

Alignment to any allele in a pangenome reference graph is as efficient as alignment to the reference allele in a linear reference sequence.
Including known genetic variation in our pangenomic reference system can thus reduce bias towards the reference allele when genotyping heterozygous variants (Figure~\ref{fig:allelebalance}).
This effect is strongest for alleles that are highly divergent from the reference, suggesting that the impact of these approaches on variant calling and genotyping will be strongest for large indels and structural variants.

\subsubsection{Sample-specific references}

Rather than implementing genotyping or variant calling over a pangenome model, the model can be used to infer a likely haploid version of a new sample's sequence, which is used as a reference genome for variant calling.
These methods show incremental improvements in accuracy over conventional methods.
\textsc{Gramtools} \cite{Maciuca_2016} and \textsc{PanVC} \cite{Valenzuela_2018} both use specialized pangenome indexes to map reads for this purpose.
\textsc{PRG} uses read $k$-mers from reads to choose likely haplotype paths through a graph \cite{dilthey2015improved}.

\subsubsection{Small variants}

In many ways, small variants stand to benefit the least from pangenomic variant calling and genotyping.
NGS read lengths are sufficient to span their entirety, and the associated variant calling algorithms are quite mature.
However, reference bias in mapping can be a source of small variant calling error, particularly for indels.

One strategy consists of realigning reads to graphs of known variation after mapping to a linear reference.
This approach was pioneered in the 1000 Genomes Project, which applied \textsc{Glia} to establish genotype likelihoods for indels and complex alleles \cite{1000_2015}.
\textsc{GraphTyper} refines this approach to achieve competitive accuracy in joint genotyping large cohorts with very low computational costs, providing improved genotyping performance at known variable sites described in its graph \cite{eggertsson2017graphtyper}.

Colored de Bruijn graphs support pangenomic, reference-free variant calling.
\textsc{Cortex} calls variants based on coverage in bubble structures in a colored de Bruijn graph, which is constructed from multiple read sets and/or reference genomes \cite{Iqbal_2012}.
\textsc{Bubbleparse} extends \textsc{Cortex}'s model to improve SNP discovery \cite{Leggett_2013}.

\subsubsection{Genotyping structural variation}

The study of structural variants (SVs)---typically defined as variants affecting at least 50 bp---has more to gain from using genome graphs.
SVs are difficult to call with NGS reads because they are large relative to the read length.
Long read sequencing does not share this difficulty, but this technology remains prohibitively expensive for population-scale studies or routine use.

\textsc{Bayestyper} \cite{sibbesen2018accurate} compares the distribution of $k$-mers from sequencing reads to the distribution of $k$-mers along paths in the graph.
It calls structural variants with high accuracy almost irrespective of the size of the variant.
However, it has a high memory footprint, and later analysis has also suggested that its reliance on long exact matches makes it fragile to breakpoint uncertainty \cite{hickey2019genotyping}.

In contrast, \textsc{Paragraph} \cite{chen2019paragraph}, \textsc{GraphTyper2} \cite{eggertsson2019graphtyper2}, and \textsc{VG call} \cite{hickey2019genotyping} use genome graphs to genotype SVs.
The largest difference is that \textsc{Paragraph} and \textsc{GraphTyper2} first map to a linear reference, then locally realign to regional graphs, whereas VG maps reads directly to a whole genome graph.
These methods all use read coverage to determine the genotype, and significantly outperform competing reference-based methods.

Some pangenomic methods have sought improved accuracy and efficiency by focusing on specific regions where reference bias makes inference challenging. 
The highly polymorphic and medically important human leukocyte antigen (HLA) genes have received an especially large amount of attention.
\textsc{HLA*LA} \cite{dilthey2019hla}, \textsc{Kourami} \cite{lee2018kourami}, and \textsc{HISAT-genotype} \cite{Kim_2019} have all demonstrated techniques for genotyping HLA genes by aligning NGS reads to a graph encoding various HLA alleles.
Their results rival gold-standard Sanger sequencing methods in accuracy.
\textsc{ExpansionHunter} \cite{dolzhenko2019expansionhunter} and \textsc{HISAT-genotype} \cite{Kim_2019} used similar methodologies to achieve comparable or better accuracy than existing methods for short tandem repeats (STRs).

\subsection{Inferring precise haplotypes from pangenomes and metagenomes}
\label{sec:haplotypes}

Current assembly approaches often produce a result that mixes both haplotypes of a diploid genome together.
This inaccurate representation has led to the development of diploid assembly methods.
\textsc{WHdenovo} \cite{garg2018graph} addresses this by using long reads to infer phase within the pangenomic space of the assembly graph.
In an alternative approach to haplotype inference, pangenomic error correction methods can be used to clean small errors from long reads, rendering them precise observations of long haplotypes \cite{Salmela2016LORMA,Rautiainen_2019b}.
These methods build an assembly from accurate reads (NGS or PacBio circular consensus), to which reads can be aligned.
The path of the alignment through the graph is taken as the corrected read.

Several methods have used pangenome graphs to support haplotype reconstruction, but in the context of a mixed population of related genomes sampled from a metagenome or quasispecies mixture.
\textsc{Mykrobe} predictor \cite{Bradley2015-kl} and \textsc{GROOT} \cite{Rowe2018-bg} use graph-based structures of bacterial genomes and gene sets to predict antibody resistance in sequencing samples.
\textsc{MetaKallisto} \cite{Schaeffer2017-fh} performs taxonomic classification and quantification of metagenomic sequencing data using a database of known sequences represented as colored de Bruijn graphs.
\textsc{Virus-VG} \cite{Baaijens2019-ng} builds a variation graph from assembled viral contigs in order to construct haplotypes and predict associated abundances in viral quasispecies from sequencing reads.
This method was later improved in \textsc{VG-Flow} \cite{Baaijens2019-ha} which can scale to much larger genomes, such as bacteria.  

\subsection{Functional pangenomics}
\label{sec:functionalgenomics}

The effects of reference bias on analyses of allele-specific protein binding and transcription have led to ample interest in pangenomic techniques.
Reference bias may also affect our ability to associate genome with phenotype in association mapping.
Here, we consider ways in which pangenomic models can improve the accuracy of such assays into genome function.

\subsubsection{CHiP-seq peak calling}

CHiP-seq data is mapped back to the reference genome in order to locate protein binding sites.
\textsc{Graph Peak Caller} is based on \textsc{VG} and is the first tool to use a genome graph for this process \cite{Grytten_2019}.
It was shown to find binding sites more enriched for known DNA binding motifs than linear methods on \emph{A. thaliana}.
It was also applied to human data to discover novel sites for enhancers \cite{groza2019personalized}.

\subsubsection{Transcriptomics}

Some transcriptomic analyses are strongly affected by reference bias.
Chief among these is allele-specific expression (ASE) \cite{Degner2009-vw,stevenson2013sources,Castel2015-ef}.
ASE analysis estimates the expression levels of genes or transcripts on each allele separately by comparing the number of RNA sequencing (RNA-seq) reads mapped to the two different alleles of heterozygous variants.
A mapping bias in favor of one of the alleles can therefore create illusory differences in expression between the alleles.
Using variation information during mapping can help ameliorate this and improve estimates of ASE \cite{Castel2015-ef,Miao2018-ps}.

The simplest approach to using variation data in mapping involves creating a personalized diploid genome or transcriptome, which is then used as the reference for a standard linear mapping method \cite{Raghupathy2018-sd}.
Methods using this approach have been shown to reduce reference bias, but they require diploid reconstruction of the genome in question.
Variant-aware mappers like \textsc{GSNAP} \cite{Wu2010-hv}, \textsc{iMapSplice} \cite{Liu_2018}, \textsc{ASElux} \cite{Miao2018-ps}, and \textsc{HISAT2} \cite{Kim_2019} remove this necessity, and have been shown to reduce reference bias at known SNVs during mapping \cite{Castel2015-ef,Liu_2018}.
Variation-aware analysis of RNA-seq data is also important for accurately analyzing highly polymorphic regions, such as HLA.
\textsc{AltHapAlignR} \cite{Lee_2018} and \textsc{HLApers} \cite{Aguiar2019-fy} compare reads against a collection of known HLA haplotypes yielding improved estimation of HLA expression.

\subsubsection{Pangenomic association studies}

Pangenome-wide association studies (PWAS) generalize the genome-wide association (GWAS) concept to pangenomes.
This new area has primarily used traditional pangenome definitions from microbiology \cite{Brynildsrud_2016}.
Recent work applies PWAS specifically to pangenome graphs.
In the frequented region (FR) technique, a syntenic region finding algorithm similar to those used in whole genome alignment (\S \ref{sec:constructionWGA}) detects regions of a compacted DBG that are shared between many individuals \cite{cleary2018exploring}, and these are employed as features in PWAS \cite{Manuweera_2019}.
When tested in 100 yeast strains, this approach marginally improves on standard GWAS techniques, but for some phenotypes provides a dramatic improvement in performance.
