\section{Applications of pangenomic models}

\subsection{Error correction}
% Robin

\subsection{Variant calling and genotyping}
% Glenn

Genome graphs can leverage known sequence variation, available in increasingly large public databases, to improve variant identification in new samples.
In general, variants are determined by finding paths through the graph that are best supported by mapped reads, and noting how the reads differ form the paths that best represent them.
Projecting these paths back to the reference yields the variants, typically in VCF format.
When only paths from graph are considered, this process is considered \emph{genotyping}.
Novel variants can be \emph{called} by augmenting the paths with edits from the reads.

\subsubsection{General tools}

GraphTyper \cite{eggertsson2017graphtyper} is a popular genome-graph-based caller.
It begins with a pangenome graph created from variants in dbSNP, and iteratively updates it with variants discovered in input reads.
Its authors showed it to be more accurate on benchmarks than top linear reference-based approaches such as GATK.
It is also practical to run on large datasets; its authors applied it to WGS data from 28,075 samples from an Icelandic population study.

A more recent approach, Graph Genome Pipeline, \cite{Rakocevic_2019}, uses a genome graph derived from the 1000 Genomes Project to improve calling accuracy.
It has been shown to be able to find variants absent from common benchmarks such as GIAB, raising the question of whether these datasets, commonly used as truth sets in the evaluation of new methods, should be updated.

\subsubsection{Approaches to challenging regions}

Very diverse genomic regions (whose diversity makes them difficult to study) and low complexity regions (which can have condensed graph representations) intuitively stand to see the most improvement when moving from a linear reference to a genome graph.
To this end, much work has been done exploring the application of graphs to the major histocompatibility complex (MHC) region on chromosome 6, which contains the human leukocyte antigen (HLA) genes.
These genes are critical components of the immune system, which produces a strong selective pressure for diversity and many dramatically distinct alleles.
Determining the HLA alleles in a sample, or HLA typing, is often of great importance, but due to high polymorphism within the MHC region, short read sequencing against a linear reference is not effective, and more costly Sanger sequencing is required.

As a potentially viable short-read-based approach, \citeauthor{dilthey2015improved} developed the Population Reference Graph (PRG) method.
The PRG is created from a multiple sequence alignment (MSA) of known haplotypes and variants spanning the MHC region.
Reads are mapped to to the graph, then used to extract a pair of ``chromotype'' paths that best represent them, which are then used linear references for standard mapping and calling tools.
An extension, PRG*HLA \cite{dilthey2018hla}, explicitly models the HLA genes, rather than the whole MHC region.
In this method, HLA types are imputed directly from the graph, rather than intermediate chromotypes, providing faster performance.
A further extension, HLA*LA \cite{dilthey2019hla}, further improves runtime and adds support for long read and assembly inputs.
Kourami \cite{lee2018kourami} is another method for HLA typing that assembles haplotypes using a graph genome as a guide, and is able to incorporate novel variation from the reads.
HLA typing with these methods using whole genome short read data has been shown to be as accurate as the conventional Sanger sequencing based approach.

For working on low-complexity, repetitive regions of the genome, ExpansionHunter \cite{dolzhenko2019expansionhunter} provides an approach using genome graph models.
It uses regular expressions are used to define variable sites of tandem repeats, which are modeled as self-loops in the graph.
Its authors show that this approach can more accurately genotype clinically relevant sites of such variation using short read data, as compared to standard variant calling pipelines.

\subsubsection{Genotyping structural variation}

The study of structural variants (SVs), mutational events of at least 50~bp in length, is also benefiting from the application of genome graphs.
SVs can be increasingly well characterized by long read sequencing, but this technology remains prohibitively expensive for population scale studies.
Known SVs can be naturally represented as genome graphs, and genotyped from short read data using a selection of tools.

Bayestyper \cite{sibbesen2018accurate} uses the distribution of kmers from sequencing reads, as well as that of the graph, to accurately identify variant haplotypes.
 

Paragraph \cite{chen2019paragraph} is a SV genotyper that operates on a genome graphs constructed from each variant in a VCF of SV calls.
For each variant, it remaps nearby reads as retrieved from a linear alignment, to the graph.
It then computes a genotype from the support of each allele's breakpoints in the graph alignment.
Paragraph was shown to outperform similar methods that rely on linear references by a wide margin.

\texttt{vg}'s SV genotyper can be run on any genome graph that contains embedded paths, which are required as the coordinate space for VCF output \cite{hickey2019genotyping}.
Unlike all other methods mentioned above, which remap a subset of relevant reads to the graph, \texttt{vg} operates directly on an alignment of the entire read set against the graph.
It uses the snarl decomposition \cite{paten2018superbubbles} to identify sites of variation in the graph, and derives haplotypes using read support.
It was shown to be much more accurate than linear reference-based approaches, and was also evaluated on graphs derived from alignments of assemblies.

\subsection{Assembly and metagenomics}
% Erik

\subsection{Transcription factor binding}
% Glenn

CHiP-seq data, reads that bind to specific transcription factors, can be mapped back to the reference genome in order to locate binding sites.
 
Graph Peak Caller is based on \texttt{vg} and is the first tool to use a genome graph for this process \cite{Grytten_2019}.
It was shown to find binding sites more enriched for known DNA binding motifs than linear methods on \emph{A.
thaliana}.
It was also applied to human data to discover novel sites for enhancers in the human genome \cite{groza2019personalized}. 

\todo[inline]{Do we need a real epigenomics section about DNA base or histone modification?}

\subsection{Transcriptomics}

Attempts to use genome graphs to analye transcriptomic sequencing data have generally received less attention than similar experiments with whole genome sequencing data.
However, reference bias is an important issue that needs to be addressed when estimating allele-specific expression (ASE \cite{Degner2009-vw,Castel2015-ef}.
ASE analysis estimates the expression levels of genes or transcripts on each allele separately, by comparing the number of RNA sequencing (RNA-seq) reads mapped to the two different alleles of heterozygous variants.
A mapping bias in favor of one of the alleles can therefore result in a false-positive difference in expression between the alleles.

Simulation-based filtering strategies, which remove variant sites or reads susceptible to biased mapping, have been proposed as a way to mitigate this, but filtering out reads can reduce sensitivity for detecting genes with allele-specific expression, and can produce lower estimated expression values overall \cite{Castel2015-ef,Van_de_Geijn2015-dz}. 
Variant-aware methods, including those based on genome graphs, do not rely on filtering and instead take the sequence difference between the two alleles into account during mapping.
The simplest approach to using variation data in mapping involves creating a personalized diploid genome or transcriptome, which is then used as the reference for a standard linear mapping method \cite{Turro2011-op,Rozowsky_2011,Bray_2016,Raghupathy2018-sd}.
However, this simple approach cannot be properly said to be variant-aware, as it has no concept of a variant.
Methods using this simple approach have been shown to reduce reference bias and improve estimation of ASE, relative to traditional single reference genome methods, but one limitation is that they require phased haplotypes for the individuals under study.

Proper variant-aware mapping methods, on the other hand, do not require that the variants are phased.
GSNAP was the first variant- and splicing-aware mapping method developed for RNA-seq data \cite{Wu2010-hv}.
It uses a kmer-based approach where both the genome and a set of SNVs are indexed using hash tables.
% We might want to add GSNAP introduction to the model section since it also works on WGS.
Current version uses a suffix array in addition to the hash table.
Alignment is performed by searching for kmers in the read that are in close proximity on the reference genome.
% Needs a much more detailed and better description.
\todo{Do we need to go into how GSNAP actually maps? Is it apprecialbly different from how non-RNA-seq graph mappers map?}
GSNAP is still competitive with regards to mapping accuracy, but it is generally much slower than contemporary mapping methods.
Although not demonstrated in the initial publication, GSNAP does reduce reference bias \cite{Castel2015-ef}.

Another variant-aware method, ASElux, uses all heterozygous exonic SNVs in an individual to create a suffix array index of the alleles and their flanking sequences \cite{Miao2018-ps}. 
This index is used to filter read pairs that does not overlap any SNVs with up to 2 mismatches allowed in the flanking regions. 
Read pairs that pass this filter are then aligned to a different suffix array index consisting of exonic and intronic regions in order to find its exact location. 
ASElux showed better reduction in reference bias and higher alignment speed compared to \todo{This said ``variation aware-mapping methods'', but that can't be right because this method is variant-aware.
Do we mean ``other variant-aware mapping methods'', and if so, which?}.

Similarly to ASElux, iMapSplice creates an index of SNV alleles and their flanking sequences \cite{Liu_2018}.
The sequences are indexed using enhanced suffix arrays and reads are mapped to both this index and the reference genome using semi-maximal prefix matching.
The authors show that iMapSplice achieves higher mapping accuracy and lower reference bias compared to both linear and other variant-aware mapping methods.

GSNAP, ASElux and iMapSplice only support SNVs and are therefore not able to reduce reference bias around indels.
HISAT2 was the first genome graph based, non-SNV-variant-aware method for splicing-aware mapping of RNA-seq data \cite{Kim_2019}. 
This method can use SNVs, deletions of any length, and insertions up to 20~bp.
Using a combination of a hierarchical graph FM index and a repeat region index (see \todo[inline]{?} section for more details) HISAT2 is able to map RNA-seq reads at an impressive speed to a genome graph containing 14.5M variant sites \todo{Does the paper give this fugure for sites or alleles?}, without sacrificing accuracy. 
The original benchmark, however, only shows results for DNA sequencing data; we are unaware of any research into how much, if at all, the variant-aware HISAT2 reduces reference bias for RNA-seq data.

Recently, the ability to create spliced variation graphs was added to the \texttt{vg} toolkit \cite{Garrison_2018}. 
In these variation graphs, known splice junctions are added as edges, similar to the addition of a deletion event, while transcripts are embedded as paths through the graph. 
This allows for existing algorithms in \texttt{vg} to also be used for variant-aware mapping of RNA-seq data. 
\texttt{vg} supports any type of variation, but its splicing-awareness is limited to only known splice junctions, and it does not support generating supplementary alignments.
Thus, reads that span a novel splice junction will only map partially.

As with whole genome DNA sequencing data, variation-aware analysis of RNA-seq data is important for getting accurate results in highly polymorphic regions, such as the HLA genes in the human major histocompatibility complex region. 
Multiple approaches have demonstrated improvements in the estimation of HLA gene expression as a result of comparing the reads against a collection of known HLA haplotypes, instead of using the linear reference.
AltHapAlignR is one such method \cite{Lee_2018}.
It first maps the RNA-seq reads to the linear reference.
Unmapped reads and reads that map to the HLA region are then collected and mapped to seven representative alternative haplotypes, followed by expression estimation in a haplotype-specific manner. 
While this approach improves estimation of ASE, it is unclear whether the method can scale to the thousands of haplotypes known in the HLA region.
In contrast, another method, HLApers, first aligns the RNA-seq reads against an index containing thousands of known HLA haplotypes using either a regular mapper or pseudoaligner \cite{Aguiar2019-fy}.
The aligned reads are then used to predict the most likely haplotype pair for each gene to which the reads are aligned against a second time and from which expression is estimated.
Using this approach it was able to align more reads than could be aligned using the linear reference, especially for genes with high sequence divergence from the reference.
However, only aligning to a particular set of HLA haplotypes might bias the results toward the haplotypes used for alignment.

Leveraging variation in the intronic regions can also be helpful when analysing RNA-seq data. 
Variation in these regions can disrupt existing splice-site motifs or create new ones, resulting in intron retention or novel splicing, respectively. 
Since the presence or absence of canonical splice-site motifs is often used as a factor when scoring alignments across splice junctions, mappers which are aware of these changes to the motifs would be expected to give more accurate mapping results for reads that could overlap them. 
Indeed, by using a personalized genome approach, \citeauthor{Stein_2015} were able to identify 506 personal splice junctions in 75 individuals, of which 437 were novel \cite{Stein_2015}. 
Later, \citeauthor{Liu_2018} used iMapSplice to find more than triple this number in the same individuals, thereby showing the importance of variant-aware analysis for reducing bias in splice junction detection \cite{Liu_2018}.

% Papers that maybe needs to be added here or under calling/genotyping:
% HLA typing from RNA-seq (Beretta \textit{et al.}, Denti \textit{et al.})


