\section{Discussion}
\label{sec:discussion}

In the near future, we expect complete, haplotype-resolved, ``telomere-to-telomere'' assemblies of large genomes to be readily obtained at low cost \cite{miga2019telomere}.
The impeding resolution of the genome assembly problem raises new issues.
To make full use of genome assemblies, we must relate them to each other.
And, to maximize their value, we must make it possible to use the prior information contained in them to guide subsequent genomic analyses.

These goals drive us to work with the pangenome implied by a collection of whole genome assemblies.
Pangenomes can be modeled as simple collections of DNA sequences, but this can obscure variation between genomes that is essential to unlocking insight into biology.
Increasingly, researchers have explored pangenome graphs which represent both sequences and variation between them.
These methods are flexible.
In representing the mutual alignment of many genomes, a graphical pangenome can contain, and relate between many linear reference systems.
They are also scalable.
Recently-developed methods support the compact storage and query of collections of tens of thousands of genomes.
And, they can improve alignment and genotyping accuracy in the context of known variation.

However, adding variation to the reference system is not without potential drawbacks.
Model construction, indexing, and alignment steps typically require more time for pangenome graphs than linear reference genomes.
Additional information can increase ambiguity, and care must be taken to build models that improve utility by including relevant variation.
Working with a graphical reference system necessitates knowledge of graph-theoretic concepts that may be unfamiliar to many biologists.
Users also must consider that, pangenome graphs are not observable in the same sense that a given genome is.
Their construction is often guided more by application than a clear ground truth.

Due to these issues, some argue that it is likely that linear genomic models will remain important into the future \cite{Sherman_2020}.
Our survey does not disagree with this possibilty.
Many of the works we have considered forsee a future in which reference systems are graphical, but only a handful (primarily those based on variation graphs) produce alignments or genotype calls in the context of a pangenome graph.
Linear or hierarchical coordinate systems for the pangenome may be preferred by the genomics community.
If so, these reference systems are likely to proliferate as we explore the pangenome of humans and other species.

Of course, a future full of many reference genomes is essentially a pangenomic one.
Whether or not the community fosters the development of standardized pangenomic reference models, the proliferation of whole genome sequences will only increase the importance of the methods that we have considered here.
Pangenome graphs provide a distributed framework which we can use to bring many reference systems into the same analytical context.
We can use them to build reference models optimized for particular research or clinical settings, potentially mixing public and private sources of data, without sacrificing our ability to relate our findings to standard reference models.
Provided their continued improvement, graphical pangenomic methods will be well-suited to the pluralistic, decentralized attributes of a future in which genomes are easily sequenced and assembled.

