\section{Discussion}
\label{sec:discussion}

The term \emph{pangenome} has previously implied the study of gene families within a given species or clade.
Technological change, in the form of improved sequencing and assembly algorithms, allows us to build pangenomes that represent collections of genomes.
Analytical methods capable of using these complete pangenome models let us study the precise evolutionary relationships between whole genomes.

New techniques are being developed to utilize this powerful prior information about genomic variability in a given species or clade.
Often, these methods rely on graph-based representations of pangenomes which capture both the sequence of and variation between represented genomes.
These methods demonstrate improved performance and accuracy when working with pangenome models relative to standard genomic ones.
They have been shown to eliminate reference bias at known variant sites, and allow the direct comparison of new data to large pangenomes.

However, it is not clear that graphical pangenome models will themselves replace linear reference systems.
Few of the methods which we have reviewed makes a strong case that the reference system itself should become a graph.
For instance, only a handful of mapping and variant calling methods (primarily those based on variation graphs) even produce alignments or genotype calls in the context of the graph, with the majority reporting them against a linear reference sequence.
In combining sequences with their alignments, graphical pangenomes confuse the traditional concepts of genome position and annotation which are essential for standard research practice.
To date, there is no widely-accepted mechanism to generalize such concepts to graphs.

We speculate that the status quo of genome positions on linear sequences may continue long into the future, even if graphical pangenome models become essential to many kinds of analysis.
On their own, pangenome graphs do not represent any directly measurable aspect of a biological system, and thus their construction and design is guided more by application than any kind of ground truth.
In this view, pangenome graphs are technical artifacts important for analysis, but may not provide a stable foundation for many ``legacy'' techniques.
However, pangenome graphs can allow us to record the direct relationship between many linear reference systems.
Thus, although their topology may not become part of the reference, these graphs allow us to harmonize diverse useful linear consensus models of the genome.

Graphical pangenomic methods, regardless of the specifics of their representation of the pangenome, aim to provide accurate and unbiased access to this collection of sequences with minimal resource costs.
These methods provide a coherent framework for thinking about the plurality of sequences in a pangenome.
They resolve fundamental problems in genomic analysis that will become ever more severe as we consider increasing numbers of fully-resolved whole genome sequences in the course of biological research.

