\section{Pangenomic data formats}
\label{sec:formats}

A number of common data formats are used to exchange pangenomic models.
Pangenomes can be stored as collections of sequences in FASTA format.
Variant calls in VCF format \cite{danecek2011variant} may be added to such a collection to describe small or structural variants found in the pangenome.
However, to exchange graphical pangenomes, the community frequently uses a subset of the GFAv1\footnote{\url{https://github.com/GFA-spec/GFA-spec/blob/master/GFA1.md}} assembly graph format.
This has the benefit of allowing the application of tools that read GFA to pangenome graphs.

GFA provides no mechanism to express read alignments.
The \textsc{VG} toolkit has developed the GAM format, which generalizes the data model of the SAM/BAM \cite{Li_2009} format to work on graphical reference systems, and this has been taken up by a number of other alignment tools.
However, it lacks a simple text-based equivalent.
The recently-proposed GAF\footnote{\url{https://github.com/lh3/gfatools/blob/master/doc/rGFA.md\#the-graph-alignment-format-gaf}} format (which generalizes PAF\footnote{\url{https://github.com/lh3/miniasm/blob/master/PAF.md}} to work on graphs) could serve this purpose.
GAF is related to rGFA\footnote{\url{https://github.com/lh3/gfatools/blob/master/doc/rGFA.md}}, which is a proposed restricted specialization of GFA for reference pangenome graphs.
