\section{Introduction}
\label{sec:intro}

A \emph{pangenome} models the full set of genomic elements in a given species or clade.
Pangenomics thus stands in contrast to reference-based genomic approaches which relate sequences to a particular consensus model of the genome (figure \ref{fig:models}).
Genomes that are reconstructed with the aid of a reference genome can appear more similar to the reference than they actually are.
Pangenomic reference systems can reduce this bias by enabling the direct relationship of new genome to all those represented in the pangenome.

Pangenomics has been important to microbiology, where genomic plasticity and diversity has made it indispensable \cite{Vernikos2015}, and it has increasingly seen application to eukaryotic genomes \cite{cao2011whole,gao2019tomato,Ou_2018}.
Standard pangenomic analyses focus on the presence or absence of genes from given strains and the determination of a core (commonly present) and accessory (frequently absent) pangenome \cite{page2015roary}.
However, they have tended to pay less attention to variation between these sequences, and they do not typically attempt to provide a precise model relating many genomes to each other at the base level.

In contrast, most ``high-throughput'' analyses of large genomes depend on comparison to a single reference genome.
However, in recent years, reduced sequencing and \emph{de novo} assembly costs have supported the discovery of significant levels of large-scale genomic variation in many eukaryotic species, including humans \cite{sudmant2015integrated,Hehir-Kwa2016-hb,chaisson2018multi,Audano_2019}, arabidopsis \cite{alonso2016arabidopsis}, brewer's yeast \cite{yue2017contrasting}, and the fruit fly \cite{chakraborty2018hidden}.
These observations have generated wide interest in extending bioinformatic operations to use a pangenomic reference model \cite{computational2016computational}.
The availability of true pangenomic references for humans \cite{Church2015-vt} and other model organisms will increasingly render the use of a single reference genome suboptimal.
However, using pangenomes effectively requires the development of new bioinformatic methods capable of constructing, querying, and operating on them.

In this review, we consider an emerging class of bioinformatic methods that let us consider genetic diversity at every stage of analysis.
We focus primarily on those methods which achieve this result by replacing linear reference genome system with a graphical, pangenomic one.
To provide background, we first consider how limitations of the current dominant genome inference paradigm motivate these methods (\S \ref{sec:resequencing}).
Then, we describe pangenomic models, with a particular focus on graphical ones (\S \ref{sec:models}).
This provides a foundation to understand how they can be constructed from sequencing data or assembled genomes (\S \ref{sec:building}).
We then survey index data structures that allow us to interact with them efficiently (\S \ref{sec:indexing}).
These index structures allow us to interrogate, visualize, and relate new information to pangenomes, such as in the process of read alignment (\S \ref{sec:relating}), yielding results in a variety of new data formats (\S \ref{sec:formats}).
Finally, we examine a number of downstream applications of these models (\S \ref{sec:applications}), and reflect on the future impact of these methods on genomics (\S \ref{sec:discussion}).

\section{Resequencing reconsidered}
\label{sec:resequencing}

Biological research frequently depends on our ability to infer the sequence level relationships between genomes.
In an earlier era, when many sequencing studies focused on small regions or single genes, it was often feasible to relate all relevant sequences to each other.
Although limited in scope, these analyses were effectively pangenomic.
They were based on many-to-many relationships between sequences, typically derived by multiple sequence alignment (MSA).
Precise MSA methods are expensive, with popular algorithms scaling cubically with the number of input sequences \cite{notredame2000t}.
It is infeasible to apply such computationally demanding methods to the data scales obtained with modern high-throughput sequencing.
Instead, high-quality genome assemblies and high-throughput sequencing have encouraged \emph{resequencing} methodologies, wherein reads from each sample are aligned to a single reference genome. % \cite{stratton2008genome}.
This approach is practical and scalable.
State of the art resequencing pipelines can jointly analyze tens of thousands of genomes \cite{Poplin_2017} at a cost per genome that is only a small fraction of the total sequencing cost.

Although efficient and conceptually simple, resequencing has a significant limitation.
Precise genomic relationships are only visible for those sequences that are similar enough to the reference genome to be alignable (figure \ref{fig:models}).
This effect is known as \emph{reference bias}.
It is strongest for structural variation or sequences that are absent from the reference system \cite{sudmant2015integrated}, but it can be relevant even for SNPs, which causes problems in allele specific expression (ASE) quantification \cite{Castel2015-ef} and in the analysis of ancient DNA \cite{zhou2017antcaller}.
Given that this bias shapes the methods that we use to establish models of the truth \cite{zook2014integrating}, it is even difficult to evaluate without paradigmatic change in our analysis techniques.

Estimates based on short read sequencing data have placed the human pangenome at between 1\% \cite{li2010building} and 10\% \cite{sherman2019assembly} larger than the GRCh38 human reference assembly.
Others have demonstrated up to several Mbp of sequence are present in each new individual and not in the reference \cite{Hehir-Kwa2016-hb,Audano_2019}.
We expect that whole genome telomere-to-telomere assemblies based on long single-molecule sequencing will provide greater insight into the extent, placement, and significance of these novel sequences \cite{miga2019telomere,Langley_2019}.
Even if we know of their existence, the reference bias inherent in resequencing will continue to limit the ability of researchers to relate new sequences to these regions.
Resequencing methods that use a pangenomic reference system should not be subject to this limitation.


