\section{Pangenomic models}
\label{sec:models}

A \emph{pangenomic model} (figure \ref{fig:models}) is a data structure that represents the genomic sequences of a population, a species, a clade, or even a metagenome \cite{computational2016computational}.
The model serves as a central coordinating entity to describe the collection of sequences and genomes in the pangenome.
Pangenomic models may take many forms, including collections of unaligned sequences or learned sequence models, but here we will focus mostly on graphical ones.

\subsection{Graphical models}

The \emph{sequence graph} serves to compress many redundant input sequences into a smaller data structure that is still representative of the full set \cite{hein1989new}.
Sequence graphs may have their nodes or edges labeled with DNA sequences, but for simplicity we will focus on the node-labeled case.
In a node-labeled sequence graph, edges indicate when concatenations of the nodes they connect occur in the sequences modeled by the graph.
Walks through a sequence graph thus include the set of sequences from which it was built.
We call them \emph{bidirected} when they represent both strands of DNA and inversions between them.
Sequence graphs were first used to represent multiple sequence alignments \cite{hein1989new,Lee_2002}.
In assembly, they have been applied to represent the full information in a set of sequencing reads (as in the \emph{string graph}) \cite{Myers_2005}, or fixed-length $k$-mers (as in the \emph{de Bruijn graph} (DBG)) \cite{Pevzner_2001}.

\emph{Genome graphs} are sequence graphs used to represent whole genome relationships \cite{Paten_2017}.
Walks through these graphs represent recombinations of the genomes included in the model.
Regions of the graph where multiple paths connect a common head and tail node, often referred to as \emph{bubbles} \cite{paten2018superbubbles}, represent variation.
\emph{Variation graphs} further structure this model by embedding the linear sequences of the pangenome as \emph{paths} \cite{Garrison_2018}\footnote{\emph{Variation graphs} are similar to \emph{variant graphs} used in textual research to model a collection of revisions of the same text \cite{schmidt2009data}}.
Paths provide a stable coordinate system that is unaffected by the manner in which the graph was built, thus supporting the coordination of positions, annotations, and alignments between variation graphs and linear reference genomes.


\section{Building a pangenome}
\label{sec:building}

Methods to construct pangenomic data structures mirror the classes of pangenomic models.
A pangenome may simply be a collection of sequences, in which case construction is similar to the genome or metagenome assembly problem.
Or, it may include information about the alignment of sequences or genomes within it.
This alignment could be compressed into a set of variants found against a set of reference sequences.
If this alignment is based on $k$-mers, then it implies a de Bruijn graph.
If it is a complete, gapped alignment, covering small and large variation, then the pangenome model can be thought of as a whole genome alignment.

\subsection{Collecting sequences}

A pangenome can be represented as a collection of sequences.
Several approaches support the construction, annotation, and interrogation of these pangenomic sequence collections.
\textsc{Panseq} \cite{Laing_2010} finds novel regions, determines the core and accessory genome, finds SNPs within the core pangenome, and then determines a subset of loci useful for molecular fingerprinting.
\textsc{PGAP} \cite{Zhao_2011} extends \textsc{Panseq}'s approach with modules for evolutionary and functional analysis, and is implemented as a single standalone executable.
Recent work has focused on scaling these techniques to ever larger genomes.
\textsc{PanTools} \cite{Sheikhizadeh_Anari_2018} detects and annotates homology groups in large collections of large genomes using a $k$-mer based approach.
Its detailed graph database model connects the panproteome defined by homology groups to genomic annotations and sequences.
\textsc{HUPAN} \citep{Duan_2019} extends the sequence collection model to human and large eukaryotic genomes, taking assembled genomes as input and finding non-reference sequences within them by comparison to a reference genome.

\subsection{Adding variation}

Rather than collecting unique sequences that represent a collection of genomes, we can consider small variants between the collection and a reference genome.
Such a model directly implies a directed acyclic graph, ordered along the reference genome, with bubbles at the sites of variation.
This pangenome construction approach is used in diverse graph genome read mappers, including \textsc{GenomeMapper} \cite{Schneeberger_2009}, Seven Bridges' \textsc{Graph Genome Pipeline}, and \cite{Rakocevic_2019}, \textsc{PanVC} \cite{Valenzuela_2018}, and \textsc{Gramtools} \cite{Maciuca_2016}.
The variation graph toolkit, specifically \textsc{VG construct} \cite{Garrison_2018}, can be applied to transform VCF files and reference sequences into genome graphs.
Some methods, like the journaled string tree \cite{Rahn_2014}, and methods based on elastic degenerate texts \cite{Bernardini_2019} such as \textsc{SOPanG} \cite{Cis_ak_2018}, transform the variant set and reference into a structure optimized for online sequence queries of the pangenome.
Deciding which variation should be added to a graph is non-trivial, and has encouraged studies of graph utility \cite{Novak_2017a} and algorithms to determine which variation is helpful \cite{Pritt_2018}.

\subsection{Colored, linked, and compacted de Bruijn graphs}

De Bruijn graph-based assemblers can be given a pangenomic quality through the addition of ``colors'' to their nodes ($k$-mers) or unitigs (unbranching components in the graph).
Each color provides a mapping between a specific biosample and a subset of the graph.
\textsc{Cortex} first demonstrated that colored DBGs could perform population scale analyses with an efficient graph implementation \cite{Iqbal_2012}.
Recent improvements to colored DBG construction, such as \textsc{Bifrost} \cite{holley2019bifrost}, allow the construction of colored DBGs from very large sequence sets (the authors build a pangenome of 118,000 \emph{Salmonella} genomes), and further support efficient updates of these pangenomic models.
The feature which makes these methods efficient, the fixed $k$ on which they are based, also limits their resolution of repetitive genomic features.
It is not feasible to build them from noisy third-generation sequencing reads.
Addressing these limitations, several methods embed linking information within the DBG that can be used to reconstruct embedded haplotypes or reads \cite{Bolger_2017,Turner_2018}.

A number of methods use compacted DBG\footnote{Compacted here means that chains of $k$-mers which contain no internal furcations are merged into a single node in the graph representation.} construction to elaborate pangenome graphs.
\textsc{SplitMEM} \cite{Marcus_2014} uses a suffix tree with suffix skips to derive the set of maximal exact matches $\geq k$ between a set of genomes.
Improving on this result in both time and space efficiency, \cite{Baier_2015} demonstrated two similar pangenome graph induction algorithms based on succinct representations of the suffix tree and the Burrows--Wheeler transform (BWT) \cite{Burrows_1994}.
\textsc{TwoPaCo} \cite{Minkin_2016} applies a probabilistic data structure to narrow the set of candidate vertexes in the compacted de Bruijn graph of a set of genomes, supporting the efficient generation of a pangenome graph from larger genomes than previous methods.

\subsection{Alignment-based sequence graphs}
\label{sec:constructionWGA}

Sequence graphs \cite{hein1989new} can be understood as representations of the mutual alignment of a set of sequences.
Alignment-based pangenome structures form the basis of a number of pangenomic approaches.
They have found use in the construction of acyclic multiple sequence alignments.
The partial order aligner \textsc{POA} \cite{Lee_2002,Grasso_2004} uses an acyclic, directed sequence graph model to build multiple sequence alignments.
\textsc{ProgressiveCactus} \cite{Armstrong_2019} produces whole genome alignments that can be rendered as sequence graphs.
\textsc{SibeliaZ} \cite{Minkin_2019} finds collinear blocks within \textsc{TwoPaCo}'s compacted DBG and applies \textsc{POA} to each to yield a whole genome alignment graph.
The multiple sequence/graph aligner (\textsc{VG msga}) \cite{Novak_2017a,Garrison_2018,Garrison_2019} generalizes the progressive approach of \textsc{POA} to build generic variation graphs that include cycles and inversions.

Not all researchers have focused on generic graphs, with several arguing that completely generic models are either computationally intractable or not relevant to important practical analyses.
\textsc{REVEAL} \cite{linthorst2015scalable} builds a pangenome graph from a syntenic set of maximal exact unique matches of decreasing size between a pair of sequences (or graphs), and later adds inversions detected by alignment of paths through bubbles in this graph.
\textsc{NovoGraph} \cite{Biederstedt2018} follows a reference-guided approach, breaking a set of genomes into syntenic alignable blocks and deriving a multiple sequence alignment for each, and yielding a VCF file as its output.
Similarly, \textsc{seq-seq-pan} \cite{Jandrasits_2018} employs existing whole genome alignment methods to find a set of locally collinear blocks, in which it compacts into a sequence graph that respects the synteny of the input genomes.
\textsc{GenGraph} \cite{Ambler_2019} realigns previously identified co-linear blocks yielding a genome graph from an MSA.

Recent, unpublished methods explore two new alternatives to alignment-based pangenome construction.
\textsc{minigraph}\footnote{\url{https://github.com/lh3/minigraph}} extends the \textsc{minimap2} \cite{Li_2018} alignment chaining model to work on graphs.
It applies this alignment model to progressively build out a pangenome graph from a series of genomes that contains large ($>$250bp) sequences that were not previously seen in other genomes.
The resulting pangenome does not contain all input sequences and variation between them, but a representative subset and large structural variants.
In contrast, \textsc{seqwish}\footnote{\url{https://github.com/ekg/seqwish}} \cite{Garrison_2019} generates the full variation graph implied by a collection of sequences and alignments between them.
The paths embedded in its output graph precisely and completely reconstruct the input sequences, while the topology of the graph describes all variants represented in the input alignments.

\subsection{Positional systems in pangenomes}
\label{sec:positions}

Reference genome sequences provide a coordinate system to catalog and exchange information about genes, protein binding sites, epigenetic profiles, variants, and homologies.
In linear references, genomic coordinates are easily interpretable, and they unambiguously indicate both the layout of the sequence and the distance between bases, but this is not the case when these coordinates are embedded within a graph \cite{Rand_2017}.

It is possible to use reference coordinates in a graphical pangenome by embedding reference sequences inside the graph and labeling graph nodes with their relative positions in these paths \cite{Garrison_2018,Garrison_2019}.
This has been extensively explored in variation graph based tools.
However, several problems remain.
The embedded coordinate systems may be incomplete, in that they may not fully cover the graph.
Also, particular graph instantiations may induce ambiguity in reference positions.
For instance, a copy number variant that is collapsed in the graph will contain multiple overlapping reference path coordinate ranges.

These limitations have driven the development of complete coordinate systems for genome graphs.
One solution is to build a hierarchy of graph components, based on a starting reference sequence, adding a new name and coordinate range for each non-reference sequence that is included \cite{Rand_2017}\footnote{\textsc{minigraph} uses a similar model}.
Another technique is to build positional systems based solely on the topology in the graph \cite{paten2018superbubbles}.
Genomic variation creates a system of nested bubble structures that can be used to spatially organize graph elements.
This approach has a rigorous, if complex, mathematical basis.
Similar decompositions of the graph topology have been used in assembly-based variant detection \cite{Iqbal_2012, Onodera_2013}.


\section{Indexing pangenomes}
\label{sec:indexing}

Index data structures for pangenome graphs support efficient random access to elements and features of the graph.
Attention must be given to ensure that these index structures do not require significant overheads relative to the information content of the graph.
Naive implementations of sequence and structural indexes of the graph can incur significant runtime and memory costs, which can become problematic as graph sizes increase.
Succinct data structures and careful encoding of these data are thus required to reliably fit large graphs into the main memory of commodity computing systems.
Particular index models lie at the core of the highest-performing graph based visualization (\S \ref{sec:viz}), read mapping (\S \ref{sec:graphmapping}), and variant calling systems (\S \ref{sec:variantcalling}).

Building \emph{text indexes} for sequences encoded in a graph is more involved than for linear references.
In graphs with regions of dense variation, the number of $k$-bp paths can grow exponentially in $k$, often rendering their complete enumeration intractable even for low values of $k$.
In order to limit the exponential growth, the index may only support relatively short query strings.
Some indexes \cite{Siren_2014} support longer queries by doing extensive preprocessing.
In others \cite{Thachuk_2013,Huang_2013,Maciuca_2016}, queries mapping to complex graph regions can be slow.
Instead of indexing the entire graph, the index may only contain $k$-mers from a simplified graph, or from specific paths of the graph.

\subsection{Indexing sequences using a graph}

The FM-index \cite{Ferragina_2005} is a text index, based on the Burrows-Wheeler transform (BWT) \cite{Burrows_1994}, that is frequently used with DNA sequences.
One variant of the FM-index, the RLCSA \cite{Maekinen_2010}, run-length encodes the BWT, allowing it to store and index a collection of similar sequences space-efficiently.
If we know a good global alignment of the sequences, we can use that information to make the index both smaller and faster \cite{Huang_2010}.
This approach was developed further in the FM-index of alignment \cite{Na_2016,Na_2018}.
Both \cite{Huang_2010}\ and \citep{Na_2016}\ use the graph induced by the alignment as a space-efficient representation of the sequences.

\subsection{Indexing acyclic graphs}

One class of graph indexing methods supports only acyclic graphs, often represented as directed acyclic graphs (DAGs).
This constraint can exist either because the acyclicity of the graph provides guarantees that simplify the problem, or because incidental features of the method's software implementation preclude use on cyclic graphs.
 
\textsc{GenomeMapper} \cite{Schneeberger_2009}, the first graph-based read aligner, was limited to such graphs.
Its indexing was also relatively simple.
\textsc{GenomeMapper} uses a simple hash-based $k$-mer index, with $k \le 13$ to limit memory usage.

\textsc{GCSA} \cite{Siren_2014} was the first attempt to generalize the BWT for graphs.
It applies a number of graph transformations that preserve the graph's sequence space while creating an unambiguous ordering for nodes.
When the graph's complexity is low, these transformations are reasonably fast and do not increase the size of the graph significantly.
However, at a certain threshold of variant density the transformed graph quickly becomes too large to handle.

\textsc{BWBBLE} \cite{Huang_2013} is a BWT-based representation for VCF-based pangenome graphs.
Simple substitutions are encoded in the sequence using IUPAC codes, and the sequence is indexed using a normal FM-index.
Because each base can be encoded using 8 different characters, the search branches at every base to cover all possible characters which admit the base searched.
In practice, most branches quickly run out of results and can be pruned from the search.
\textsc{BWBBLE} represents insertions and deletions with extra sequences, including a given amount of context around the variant.
The length of this context is an effective upper bound for query length.

The \textsc{vBWT} \cite{Maciuca_2016} took another approach to using the BWT for indexing VCF-based pangenome graphs.
It encodes variants as \texttt{(ref|alt1|alt2|\dots)} in the sequence.
When the search encounters a variant, it must branch to handle each allele separately.
Both \textsc{BWBBLE} and \textsc{vBWT} trade faster index construction for slower queries.
However, a combination of IUPAC codes for substitutions, the \textsc{vBWT} approach for other variants, and a $k$-mer index for matching the first 5--10 bases, is faster than either of the originals \cite{Buechler_2019}.

\subsection{General graphs}

Some text indexes are based on Lempel--Ziv parsing or context-free grammars.
These indexes first find partial matches between the query string and the indexed phrases and then combine the partial matches into full matches using two-dimensional range queries.
In the hypertext index \cite{Thachuk_2013}, each node is a separate phrase.
Queries mapping to a single node or crossing a single edge can be matched efficiently, while finding mappings to complex graph regions can be slow.

Techniques similar to the \textsc{GCSA} can be used to represent de~Bruijn graphs \cite{Bowe_2012}.
If the graph transformations used in \textsc{GCSA} construction are stopped after $i$ steps, the resulting graph is equivalent to an order-$2^{i}$ de~Bruijn graph.
This de~Bruijn graph can be used to approximate the original graph.
Queries of this index yield no false negatives, but matches longer than $2^{i}$ may be false positives.
By using this approach, \textsc{GCSA2} \cite{Siren_2017} attempts to support fast queries in arbitrary graphs.

\textsc{GCSA2} faces the same issues with complex graphs as \textsc{GCSA}.
In practice, most graphs must be simplified before they can be indexed.
Typical simplifications include removing high-degree nodes and complex regions from the graph and replacing them with the reference sequence.
If a collection of haplotypes is available, the removed regions can be replaced with new subgraphs that contain separate paths for each distinct local haplotype \cite{Siren_2019}.
This way, the index contains all $k$-mers from the haplotypes, while usually missing some $k$-mers from their recombinations.

\subsection{Indexing graphs using sequences}

Instead of attempting to index the entire graph, it is often sufficient to index only selected paths in it.
\textsc{CHOP} \cite{Mokveld_2018} takes the paths corresponding to haplotypes and breaks them into smaller pieces.
The distinct pieces form an artificial linear reference, which can be used with any read aligner.
The process guarantees that any substring of the haplotypes of length $k$ is also a substring of one of the pieces.
As with \textsc{BWBBLE}, $k$ represents an effective upper bound for query length.

The \textsc{PSI} \cite{Ghaffaari_2019} follows a similar approach with artificial paths.
Instead of using haplotypes, \textsc{PSI} uses a greedy algorithm to find a set of paths that covers as many $k$~bp windows in the graph as possible.
When a fully sensitive index is needed, \textsc{PSI} can reverse the role of the query strings and the graph.
While complex graph regions may contain an excessive number of $k$-mers, the reads mapping to them only contain a limited number of $k$-mers.
By indexing a batch of reads and searching for the complex regions in that index, all mappings of the query strings to the graph can be found with reasonable resources.

\subsection{Indexing haplotypes and genomes in variation graphs}

Haplotypes in related individuals are typically highly similar, and thus compressible.
A series of results lead from a compact haplotype index for biallelic SNPs to a generic haplotype index for complex variation graphs.
The positional BWT (\textsc{PBWT}) \cite{Durbin_2014} provides an efficient compressed representation of a set of haplotypes over biallelic variable sites.
The \textsc{PBWT}, like the BWT, supports efficient haplotype matching queries such as maximal exact match finding.
Later work \cite{Gagie_2017} showed the PBWT to be equivalent to the wavelet matrix \cite{Claude_2015}, which is to date the most efficient known encoding of strings with large alphabets supporting a variety of important random access queries.
The graph positional BWT (\textsc{gPBWT}) \cite{Novak_2017} extends the PBWT model to haplotype walks embedded in complex sequence graphs.
The graph BWT (\textsc{GBWT}) \cite{Siren_2019} builds on a number of assumptions that tend to hold for sequence variation graphs to improve runtime and memory costs relative to the \textsc{gPBWT}.
Provided the variation graph on which they are built encodes the full complement of variation in the pangenome, both the \textsc{gPBWT} and \textsc{GBWT} are excellent, and in principle lossless, compressors of genomes.
A \textsc{GBWT} for the 5008 haplotypes in the 1000 Genomes Project required 14.6 GB, or $\approx$1 bit per 1 kbp of encoded genomic sequence.



