\section{Building a pangenome}

Methods to construct pangenomic data structures mirror the classes of pangenomic models.
A pangenome may simply be a collection of sequences, in which case construction is similar to the genome or metagenome assembly problem.
Or, it may include information about the alignment of sequences or genomes within it.
This alignment could be compressed into a set of variants found against a set of reference sequences.
If this alignment is based on $k$-mers, then it implies a de Bruijn graph.
If the alignments is a complete, gapped alignment, covering small and large variation, then the pangenome model can be thought of as a whole genome alignment.

\subsection{Collecting sequences}

% oldschool pan-genomics
A pangenome can be represented as a collection of sequences.
Determining which sequences provide a compact represenation of a pangenome is non-trivial, and a number of software systems have been developed to support this process.

Panseq \cite{Laing_2010} finds novel regions, determines the core and accessory genome, finds SNPs within the core pangenome, and then determines a subset of loci useful for molecular fingerprinting.
PGAP \cite{Zhao_2011} extends Panseq's approach with modules for evolutionary and functional analysis, and is implemented as a single standalone executable.
% pangenome using a collection of sequences model
Recent work has focused on scaling these techniques to ever larger genomes.
PanTools \cite{Sheikhizadeh_Anari_2018} detects and annotates homology groups in large collections of large genomes using a $k$-mer based approach.
Its detailed graph database model connects the pan-proteome defined by homology groups to genomic annotations and sequences.
HUPAN \citep{Duan_2019} extends the sequence collection model to human and large eukaryotic genomes, taking assembled genomes as input and finding non-reference sequences within them by comparison to a reference genome.
Despite the availability of these toolkits, it appears that many published pangenomes are established with ad-hoc methods to determine and annotate novel sequences.

%These approaches typically apply assembly methods to determine contigs representative of the pangenome, and then align data from individual biosamples back to them to determine their significicance.
%Standard alignment and variant calling methods typically work without issue on a pangenome stored as a collection of sequences, and can be applied in custom pipelines tailored to particular species contexts.

\subsection{Adding variation}

Rather than collecting unique sequences that represent a collection of genomes, we can consider small variants between the collection and a reference genome.
Such a model directly implies a directed acyclic graph, ordered along the reference genome, with bubbles at the sites of variation.
It has been key to the first wave of precision pangenomic methods.
%These ``reference and variant'' models cannot compactly represent complex structural variation (such as inversions and copy number amplifications).
%But this has been largely irrelevant to the field.
%The low-cost short read sequencing that is used for pangenomic surveys does not easily support their detection, and so much of the population scale pangenomic information made available in the last decade has been for small variants that comfortably fit into this model.

%A number of methods have been developed around this pangenomic model, but typically they use a specialized or limited version of the model as in internal data structure to drive improvements in read alignment and genome inference.
These methods typically use a specialized or restricted version of the ``reference plus variation'' pangenome graph.
GenomeMapper \cite{Schneeberger_2009} first introduced the linear \emph{genome graph} a reference system for resequencing.
To facilitate its alignment model (which has exponential costs in the number of traversed bubbles), it breaks the graph into blocks of a fixed reference length (typically 256bp).
The journaled string tree \cite{Rahn_2014} supports time and space-efficient windowed traversal over a large collection of similar sequences represented as phased series of small variants relative to a reference.
The Population Reference Graph (PRG) model \cite{dilthey2015improved}, which was validated primarily in the human MHC region, takes a more curated approach to graph construction.
Over multiple phases it builds an acyclic genome graph from $k$-mer based alignment of long haplotypes, additional gene alleles, and finally small variants.
To support its alignment (which has exponential costs in the number of bubbles) and indexing systems, gramtools \cite{Maciuca_2016} combines variants found within a given window length ($\approx$30bp) into a single locus.
PanVC \cite{Valenzuela_2018} employs a gapped MSA as its pangenomic model, inferring underlying recombinant haplotype that may then be used as a reference for alignment and variant calling with standard tools.
Seven Bridges' Graph Genome Pipeline \cite{Rakocevic_2019} was vetted on a graph with common variation and indels, which is used primarily to improve read alignment, with alignments reported against the linear reference.
The variation graph toolkit, specifically \textit{vg construct} \cite{Garrison_2018}, can be applied to transform VCF files and reference sequences into genome graphs, which can be rendered into GFA format and are subsequently used by other tools for alignment and genome annotation.
Deciding which variation should be added to a graph is non-trivial, and has encouraged studies of graph utility \cite{Novak_2017a} and algorithms to determine which variation is helpful \cite{Pritt_2018}.

%We observe out that most of these methods (excluding GenomeMapper and vg) produce reference-relative alignments or recombinant haplotypes inferred by matching a read set against their pangenomic model.
%Thus, their models are not pangenome graph references in the same sense as linear genome references.

%\subsection{Measuring model utility}

%Briefly discuss the ``Genome graphs'' paper, \cite{Novak_2017a}, findings in the vg paper \cite{Garrison_2018}, and 
%FORGe \cite{Pritt_2018}.

\subsection{Colored, linked, and compacted de Bruijn graphs}

De Bruijn graph based assemblers can be given a pangenomic quality through the addition of ``colors'' to their nodes ($k$-mers) or unitigs (unbranching components in the graph).
Each color provides a mapping between a specific biosample and a subset of the graph.
Efficiently storing such colors allows for population scale variant calling within the context of the graph, as was first demonstrated with Cortex assembler \cite{Iqbal_2012}.
Recent improvements to colored DBG construction, such as Bifrost \cite{holley2019bifrost}, allow the construction of cDBGs from very large sequence sets (the authors build a pangenome of 118,000 \emph{Salmonella} genomes), and further support efficient update of these pangenomic models.
The feature which makes these methods efficient, the fixed $k$ on which they are based, also limits their resolution of repetitive genomic features.
It is not feasible to build them from noisy third-generation sequencing reads.
Addressing these limitations, several methods embed linking information within the DBG that can be used to reconstruct embedded haplotypes or reads \cite{Bolger_2017,Turner_2018}.

A number of methods use compacted DBG\footnote{Compacted here means that chains of $k$-mers which contain no internal furcations are merged into a single node in the graph represenation.} construction to elaborate pangenome graphs.
SplitMEM \cite{Marcus_2014} uses a suffix tree with suffix skips to derive the set of maximal exact matches $\geq k$ between a set of genomes.
Improving on this result in both time and space efficiency, \cite{Baier_2015} demonstrated two similar pangenome graph induction algorithms based on succinct representations of the suffix tree and the Burrows--Wheeler transform (BWT) \cite{Burrows_1994}.
TwoPaCo \cite{Minkin_2016} applies a probabilistic data structure to narrow the set of candidate verticies in the compacted de Bruijn graph of a set of genomes, supporting the efficient generation of a pangenome graph from larger genomes than previous methods.
Thus far, no implementations of methods like these which record each genome in the graph have matched the scalability of ``colored'' DBGs.

\subsection{Alignment-based sequence graphs}

The sequence graph model of multiple alignment has long served as a mechanism to reduce the complexity of many to many relationships required to develop parsimonious models for collections of related genomic sequences.
As observed in its first description \cite{hein1989new}, by compressing ancestrally equivalent sequences into single entities in the graph, this model can accelerate alignment and other inference methods based on a pangenomic collection of sequences.
This model was later used as the basis for the partial order alignment (POA) sequence to graph alignment and graph to graph aligner PO-POA \cite{Lee_2002,Grasso_2004}.
It was generalized to represent \emph{whole genome alignments} by allowing cyclic and inverting structures in Cactus \cite{Paten_2011}, yielding superb results in evalution during the Alignathon competition \cite{earl2014alignathon}.
The multiple sequence/graph aligner (vg msga) \cite{Novak_2017a,Garrison_2018,Garrison_2019} generalizes the progressive alignment model of POA to work on generic variation graphs, including cycles and inversions.
Although it requires linear time in the number of sequences, it suffers from high memory and runtime for large sequences due to the expense of vg's indexing and alignment algorithms.
It has found use where the number of sequences is small, or sequences are short, such as in viral pangenomic studies \cite{Baaijens2019-ng}

However, not all researchers have focused on generic graphs, with several arguing that complex structures are not relevant to many kinds of analyses, or computationally intractable. % to consider in a primary model of whole genome alignment.
The recursive hierarchical alignment model in REVEAL \cite{linthorst2015scalable} finds a set of maximal exact unique matches of decreasing size between a pair of sequences (or graphs), and uses these and the flanking un-matched sequences to build a pangenome graph.
%By applying the same alignment algorithm to the sequences in the nodes of the graph, REVEAL supports the alignment of sequences to graphs, or graphs to graphs, supporting the construction of pangenome graphs for human sized genomes.
Its model is initially linear, but it can include inversions detected post-hoc by alignment of alternative paths through bubbles.
The linear alignment-based pangenome model is further explored by NovoGraph \cite{Biederstedt2018}, seq-seq-pan \cite{Jandrasits_2018}, and GenGraph \cite{Ambler_2019}.
NovoGraph follows a reference-guided approach, breaking a set of genomes into syntenic alignable blocks and deriving a multiple sequence alignment for each, and yielding a VCF file as its output.
Similarly, seq-seq-pan employs existing whole genome alignment methods to find a set of locally collinear blocks, in which it compacts into a sequence graph that respects the synteny of the its input geomes.
GenGraph realigns previously identified co-linear blocks yielding a genome graph from an MSA.

Not all methods generate a pangenome model composed of all input sequences without bias.
minigraph\footnote{\url{https://github.com/lh3/minigraph}} extends the minimap2 \cite{Li_2018} alignment chaining model to work on acyclic, non-inverting sequence graphs.
It can be used to construct a graph based on an initial linear reference by progressively augmenting the graph with new sequences that are not covered by minimizer chains matching existing components of the graph.
Although it does not embed the sequences (other than the first reference) which it was built from, the resulting graph should include representative versions of each genomic locus in the pangenome.
It is thus useful for developing a single hierarchical coordinate system for a pangenome, or to extract a set of representative sequences from a collection of assembled genomes.

Unlike most other pangenome graph construction methods, seqwish\footnote{\url{https://github.com/ekg/seqwish}} \cite{Garrison_2019} is a simplistic graph inducer.
It generates the variation graph implied by a collection of sequences and alignments between them, using disk-backed implicit interval trees\footnote{See \url{https://github.com/lh3/cgranges} and \url{https://github.com/mlin/iitii}} to scale to large pangenome sizes.
The resulting graph depends only on the input alignment set and sequences, allowing external configuration of the graph via changes in the way the alignments are generated or filtered.

%It maps a set of sequences and alignments between them to the variation graph implied by the 
%is a variation graph induction algorithm.
%It 


% syntenic by design
% very similar to novograph, linear, based on whole genome alignments, generates locally collinear blocks
% reference guided, linear alignments, MSA based, makes a linear variation graph, outputs VCF

%Several unpublished methods extend previous approaches to
%Most alignment based pangenome graph construction methods present a monolithic 




\subsection{Positional systems in pangenomes}

%One of the most important uses of conventional reference genomes is to provide a positional system for genomic analyses.
%Researchers report many genomic features and annotations relative to positions on a reference genomes: variants, genes, binding sites, epigenetic profiles, homologies, and more.
Reference genome sequences provide a coordinating system to catalog and exchange information about genes, protein binding sites, epigenetic profiles, variants, and homologies.
In linear references, genomic coordinates are easily interpretable, and they unambiguously indicate both the layout of the sequence and the distance between bases, but this is not the case when these coordinates are embedded within a graph \cite{Rand_2017}.

It is possible to use reference coordinates in a graphical pangenome by embedding reference sequences inside the graph and labeling graph nodes with their relative positions in these paths \cite{Garrison_2018,Garrison_2019}.
This has been extensively explored in variation graph based tools.
However, several problems remain.
The embedded coordinate systems may be incomplete, in that they may not fully cover the graph.
Also, particular graph instantiations may induce ambiguity in reference positions.
For instance, a copy number variant that is collapsed in the graph will have multiple reference path coordinates inside of it.

These limitations have driven the development of complete coordinate systems for genome graphs.
One solution is to build a hierarchy of graph components, based on a starting reference sequence, adding a new name and coordinate range for each non-reference sequence that is included \cite{Rand_2017}\footnote{minigraph uses a similar model}.
%Some, like \citeauthor{Rand_2017} \citep{Rand_2017},  propose some coordinate systems tailored for the GRCh38 alternative contigs that provide some (but not all) of the positive properties associated with linear coordinates.
Another technique is to build coordinates based solely on the topology in the graph \cite{paten2018superbubbles}.
Genomic variation, represented in the form of a hierarchy of bubbles in the graph, is used to derive distance estimators.
This decomposition of the graph has seen previous use in assembly based variant detection \cite{Iqbal_2012, Onodera_2013}, and has a rigorous, if complex, mathematical basis.


\subsection{Pangenomic data formats}

A number of common data formats are used to exchange pangenomic models.
Pangenomes can be stored as collections of sequences in FASTA format.
Variant calls in VCF format may be added to such a collection to describe small or structural variants found in the pangenome.
However, to exchange graphical pangenomes, the community frequently uses a subset of the GFAv1\footnote{\url{http://gfa-spec.github.io/GFA-spec/}} assembly graph format.
This has the benefit of allowing the application of tools that read GFA to pangenome graphs.

GFA provides no mechanism to express read alignments.
The vg toolkit has developed the GAM format, which generalizes the data model of the SAM/BAM \cite{Li_2009} format to work on graphical reference systems, and this has been taken up by a number of other alignment tools.
However, it lacks a simple text-based equivalent.
The recently-proposed GAF\footnote{\url{https://github.com/lh3/gfatools/blob/master/doc/rGFA.md#the-graph-alignment-format-gaf}} format (which generalizes PAF\footnote{\url{https://github.com/lh3/miniasm/blob/master/PAF.md}} to work on graphs) could serve this purpose.
GAF is related to rGFA\footnote{\url{https://github.com/lh3/gfatools/blob/master/doc/rGFA.md}}, which is a proposed restricted specialization of GFA for reference graphs.


\section{Indexing pangenomes} % was indexing genome graphs

%\todo{JAS: Is a bit more introduction needed explaining what an index can be used for?}
As pangenome graphs become large, providing random access to their topology, sequences, and embedded haplotypes becomes non-trivial.
Succinct data strctures and careful encoding of these data are required to reliably fit large graphs into the main memory of commodity computing systems.
These data structures act as self indexes of the graph, providing random access to the elements of the graph in minimial space.

A \emph{text index} maps query strings to their occurrences in the indexed text.
The occurrences are typically reported as a list of starting positions, from which one can easily determine the substrings matching the query.

Indexing the sequences encoded in a graph is more involved.
The number of $k$~bp paths often grows exponentially in $k$.
Indexing or even enumerating all $k$-mers in the graph may be thus be infeasible for reasonable values of $k$, and if the starting position of the occurrence reported by the index is in a complex region of the graph, there may be an exponential number of $k$~bp paths to investigate.

In practice, indexes for genome graphs must make trade-offs not encountered in text indexes.
In order to limit the exponential growth, the index may only support relatively short query strings.
Some indexes (e.g.\ \cite{Siren_2014}) support longer queries by doing extensive preprocessing.
In other indexes (e.g.\ \cite{Thachuk_2013,Huang_2013,Maciuca_2016}), queries mapping to complex graph regions can be slow.
Instead of indexing the entire graph, the index may only contain $k$-mers from a simplified graph, or from specific paths of the graph.
While finding the path matching the query may be expensive in some cases, indexes typically save space by only reporting the starting position of the match.

\subsection{Indexing sequences using a graph}

The FM-index \cite{Ferragina_2005} is a text index, based on the Burrows--Wheeler transform (BWT) \cite{Burrows_1994}, that is frequently used with DNA sequences.
One variant of the FM-index, the RLCSA \cite{Maekinen_2010}, run-length encodes the BWT, allowing it to store and index a collection of similar sequences space-efficiently.
\citeauthor{Huang_2010}\ \cite{Huang_2010} observed that if we know a good global alignment of the sequences, we can use that information to make the index both smaller and faster.
\citeauthor{Na_2016}\ \cite{Na_2016,Na_2018} developed this approach further in their FM-index of alignment.
While the articles do not mention it, both \citeauthor{Huang_2010}\ and \citeauthor{Na_2016}\ use the graph induced by the alignment as a space-efficient representation of the sequences.

\subsection{Indexing acyclic graphs}

One class of graph indexing methods supports only acyclic graphs, often represented as directed acyclic graphs (DAGs).
This constraint can exist either because the acyclicity of the graph provides guarantees that simplify the problem, or because incedental features of the method's software implementation preclude use on cyclic graphs.
Pangenome graphs built from VCFs represent each contig as a mostly linear run of graph nodes, branching to allow only for substitutions and possibly insertions and deletions.
%One common class of DAGs, here called \emph{VCF graphs}, 
%\todo{``VCF graph'' is a placeholder; is there a better term for this shape?}

GenomeMapper \cite{Schneeberger_2009} was the first graph-based read aligner.
It builds a VCF graph from a reference sequence and a set of variants.
To index the graph, GenomeMapper uses a simple hash-based $k$-mer index, with $k \le 13$ to limit memory usage.

GCSA \cite{Siren_2014} was the first attempt to use the BWT with graphs.
It either takes a prebuilt DAG or generates one from a multiple alignment of sequences.
GCSA then applies a number of graph transformations that preserve the language, or set of end-to-end sequences, until the nodes can be unambiguously sorted by the labels of the paths starting from the node.
When the complexity of the graph is sufficiently low, these transformations are reasonably fast and do not increase the size of the graph significantly.
However, a phase transition happens at a certain threshold of variant density, above which the transformed graph quickly becomes too large to handle.
%\todo{JME: threshold of what? variant density?}

BWBBLE \cite{Huang_2013} is a BWT-based representation for VCF-based pangenome graphs.
Simple substitutions are encoded in the sequence using IUPAC codes, and the sequence is indexed using a normal FM-index.
Because each base can be encoded using 8 different characters, the search branches at every base to cover all possible characters which admit the base searched.
In practice, most branches quickly run out of results and can be pruned from the search.
Insertions and deletions produce separate sequences, with selected amount of context around the variant.
The length of this context is an effective upper bound for query length.

The vBWT \cite{Maciuca_2016} took another approach to using the BWT for indexing VCF-based pangenome graphs.
It encodes variants as \texttt{(ref|alt1|alt2|\dots)} in the sequence.
When the search encounters a variant, it must branch to handle each allele separately.
Both BWBBLE and vBWT trade faster index construction for slower queries.
However, a combination of IUPAC codes for substitutions, the vBWT approach for other variants, and a $k$-mer index for matching the first 5--10 bases, is faster than either of the originals \cite{Buechler_2019}.

\subsection{General graphs}

Some text indexes are based on Lempel--Ziv parsing or context-free grammars.
These indexes first find partial matches between the query string and the indexed phrases and then combine the partial matches into full matches using two-dimensional range queries.
In the hypertext index \cite{Thachuk_2013}, each node is a separate phrase.
Queries mapping to a single node or crossing a single edge can be matched efficiently, while finding mappings to complex graph regions can be slow.

\citeauthor{Bowe_2012}\ \cite{Bowe_2012} used techniques similar to GCSA for representing de~Bruijn graphs.
If the graph transformations used in GCSA construction are stopped after $i$ steps, the resulting graph is equivalent to an order-$2^{i}$ de~Bruijn graph.
This de~Bruijn graph can be used to approximate the original graph.
There are no false negatives, but matches longer than $2^{i}$ may be false positives.
By using this approach, GCSA2 \cite{Siren_2017} attempts to support fast queries in arbitrary graphs.
%\todo{JAS: Is the above paragraph too complicated for non-expert readers. SH: I had to read it 2 times to get it. Maybe just say that "If GCSA construction is stopped within / at a  transformation step, the resulting graph is equivalent to a de~Bruijn graph". Can we be so "inexact"?}
% EG: this is pretty precise stuff, so I'm leaving it

GCSA2 faces the same issues with complex graphs as GCSA.
In practice, most graphs must be simplified before they can be indexed.
Typical simplifications include removing high-degree nodes and complex regions from the graph and replacing them with the reference sequence.
If a collection of haplotypes is available, the removed regions can be replaced with new subgraphs that contain separate paths for each distinct local haplotype \cite{Siren_2019}.
This way, the index contains all $k$-mers from the haplotypes, while usually missing $k$-mers from their recombinations.

\subsection{Indexing graphs using sequences}

Instead of attempting to index the entire graph, it is often sufficient to index only selected paths in it.
CHOP \cite{Mokveld_2018} takes the paths corresponding to haplotypes and breaks them into smaller pieces.
The distinct pieces form an artificial linear reference, which can be used with any read aligner.
The process guarantees that any substring of the haplotypes of length $k$ is also a substring of one of the pieces.
As with BWBBLE, $k$ represents an effective upper bound for query length.

The ``Pan-genome [sic] Seed Index'' (PSI) \cite{Ghaffaari_2019} follows a similar approach with artificial paths.
Instead of using haplotypes, PSI uses a greedy algorithm to find a set of paths that covers as many $k$~bp windows in the graph as possible.
An index using these paths alone already works well in practice.

When a fully sensitive index is needed, PSI can reverse the role of the query strings and the graph.
While complex graph regions may contain an excessive number of $k$-mers, the reads mapping to them only contain a limited number of $k$-mers.
By indexing a batch of reads and searching for the complex regions in that index, all mappings of the query strings to the graph can be found with reasonable resources.

\subsection{Indexing haplotypes and genomes in variation graphs}

Storing a set of genome embedded in a variation graph can allow us to develop positional system over the graph, which is important for practical use.
However, it is not scalable to do so for every genome or haplotype which we have in a large pangenome.
Recording each path as a series of steps through the graph would require space proportional to the length of all paths, which quickly becomes untenable.
To store many haplotypes, we must exploit the repetitiveness of the haplotype set.
This is feasible because haplotypes from individuals of the same species tend to be highly similar.

A series of results lead from a compact haplotype index for biallelic SNPs to a generic haplotype index for complex variation graphs.
The positional BWT (PBWT) \cite{Durbin_2014} provides an efficient compressed representation of a set of haplotypes over biallelic variable sites.
It does so by recording a series of permutations of haplotypes over the sites that, when executed, maintain the invariant that the haplotypes are sorted by their prefixes leading up to the site.
The PBWT, like the BWT, supports efficient haplotype matching queries such as maximal exact match finding.
Later work \cite{Gagie_2017} shows the PBWT to be equivalent to the wavelet matrix \cite{Claude_2015}, which is to date the most-effcient known encoding of strings with large alphabets supporting a variety of important random access queries.
The graph positional BWT (gPBWT) \cite{Novak_2017} extends the PBWT model to haplotype walks embedded in complex sequence graphs.
The graph BWT (GBWT) \cite{siren2018haplotype} builds on a number of assumptions that tend to hold for sequence variation graphs to improve runtime and memory costs relative to the gPBWT.
Provided the variation graph on which they are built encodes the full complement of variation in the pangenome, both the gPBWT and GBWT are excellent, and in principle lossless, compressors of genomes.
A GBWT for the 5008 haplotypes in the 1000 Genomes Project required 14.6 GB, or $\approx$1 bit per 1 kbp of encoded sequence.



