% Heading 3
\section{Relating new information to the model}
This is dummy text. This is dummy text. This is dummy text. This is dummy text. 


\subsection{Variation Graph Mappers}


There are several mapping algorithms for variation graphs which extend concepts from linear mapping algorithms into a graph space.
The most common framework for variation graph mapping algorithms is a seed-and-extend approach that parallels similar algorithms for linear genomes.
Generally, these algorithms first index the kmers in the graph using a hash-based index. 
The index is used to find short seed alignments between kmers in the read and on the graph and the seeds are clustered by their position on the graph.
Alignments in the clusters are then extended using dynamic programming to get a final alignment.

GenomeMapper is the first(?) algorithm to align a sequence to multiple genomes. 
It does not have clustering step but merges adjacent seeds to get nearly identical maximal substrings (NIMS) that are then extended using k-banded alignment.

vg uses GCSA2 for seeding to find super-maximal exact matches (SMEM), exact matches between the read and graph that cannot be extended in either direction and which have no other extensions in the graph. 
SMEMs are clustered using a path-based distance estimation and chained together using a Markov model that favors long SMEMs and short gaps between SMEMs.
From the clusters, subgraphs are extracted and transformed into DAGs and banded dynamic programming and GSSW is used to get an alignment.

V-MAP is a mapping algorithm that finds a small subgraph that a read can be aligned to using existing mappers. 
It uses an embedding of the graph based on the distance in an undirected view of the graph from a source vertex to each graph vertex.
A seeding step is done on the embedding and the interval on the embedding that contains the most seeds is used to induce a subgraph

GraphAligner is an alignment algorithm for long reads.
It can use three different methods for seeding: minimizers, maximal unique matches (MUMs) and maximal exact matches (MEMs)
GraphAligner's extension phase uses a bitvector banded dynamic programming algorithm that was adapted from a linear alignment algorithm.

%Seven Bridges also has a seed-and-extend algorithm but I don't know how it works

HISAT2 DNA and RNA aligner that doesn't use a kmer based index.
Instead, it uses an adaptation of the Ferragina Manzini (FM) index, a hierarchical graph FM index (HGFM).
The HGFM is comprised of two FM indexes: a global index of the entire genome and local indexes of smaller portions of the genome and their variants.
Repeat sequences are combined into a separate index, keeping only one copy of each repeat so sequences in repetitive regions are only aligned once.
 

Each of these graph alignment tools have demonstrated an increase in sensitivity over mapping to a standard linear reference.


