\section{Relating new information to the pangenome}
\label{sec:relating}

Here we focus on two major avenues to interact with pangenome graphs.
By \emph{visualizing} these graphs, we can gain insight into the relationship between genomes, learning about variation small and large.
Pangenome models are often used as a reference system for \emph{sequence alignment}, which enables the relation of a new biosample to the pangenome to support a wide variety of downstream applications.

\subsection{Visualization}
\label{sec:viz}


The visualization of pangenome graphs presents significant challenges that have proven to be difficult to resolve in a single system.
We present an overview of such tools in Table~\ref{table:Visualization_Features}, while Figure~\ref{fig:visualization} compares different visualization methods applied to the same graph.

Complex, nonlinear graph structures are difficult to present in a convenient number of dimensions.
Traditional genome browsers organize information in a 2D rectangle, with a linear reference's sequence space providing the horizontal dimension, and a stack of annotations plotted in the vertical dimension \cite{Haeussler_2018}.
When visualizing a graph, there are two basic approaches: attempt to mimic linear browsers and compress the graph into one dimension, or give the graph two dimensions and represent annotations using interactivity, labels, or color.
Either option is challenging, as pangenome graphs are often \emph{nonplanar}, requiring at least three dimensions to draw without overlapping lines.
Force-directed layout, an optimization-based graph layout method, is a popular approach in the second category, but displaying annotations in this layout is challenging.

Visualization tools be categorized by the kinds of graphs they are intended to display, in addition to whether they can linearize a graph.
Many tools, including \textsc{Bandage} \citep{Wick_2015} and \textsc{GfaViz} \cite{Gonnella_2018}, are designed for interpreting assembly graphs. 
These tools tend to focus on displaying the overall, rather than base-level, structure of the graph, and have no visual features to reflect the pangenomic aspect of the graph.

Tools initially designed for variation graph visualization, on the other hand, tend to focus more on base-level structure and pangenomic relationships.
For example, \textsc{Sequence Tube Map} \cite{Beyer_2019} displays precise base-scale variation, haplotypes, and short read mapping locations, using a visual language inspired by transit system maps.

There is a difference in scale when moving from an assembly or scaffold graph to a comprehensive variation graph of a species pangenome, and this scale difference also separates different visualization tools.
High-level assembly-graph tools like \textsc{Assembly Graph Browser} (\textsc{AGB}) \citep{Mikheenko_2019} struggle to show fine details in the graph, while low-level variation graph tools like \textsc{VG view} \cite{Garrison_2018} and \textsc{Sequence Tube Map}, cannot scale to large graphs.
One tool which works well at both large scale and high detail is \textsc{MoMI-G} \cite{yokoyama_momi-g:_2019}. 
Designed as a ``multi-scale'' graph browser, it presents both a \textsc{Sequence Tube Map} of base-level differences and a \textsc{Circos} \cite{Krzywinski_2009_Circos} plot of chromosomal-scale connections, and can uniquely visualize long reads in the context of pangenomic haplotypes \cite{yokoyama_momi-g:_2019}.
\textsc{ODGI viz}\footnote{\url{https://github.com/vgteam/odgi}}, uses binning and direct rendering to a raster image to generate visualizations representing gigabase scale pangenomes.
The approach is a rasterized version of the linear layout technique of \textsc{VG viz} \citep{Garrison_2019}.

Interactively visualizing human-genome-scale, human-pangenome-detail graphs coherently across zoom levels remains an open problem.


\subsection{Graph alignment algorithms}
\label{sec:graphalignment}

Sequence comparison is at the core of many genomic analyses, and sequence alignment is the essential method for doing so. 
Classic algorithms like Smith-Waterman \cite{Smith_1981} do not directly apply to genome graphs.
However, the recurrence relations that drive their scoring and traceback routines can be extended to allow the alignment of sequences to acyclic sequence graphs, as popularized by the partial order aligner \textsc{POA} \cite{Lee_2002}.
Further generalizations support the alignment of sequences graphs to sequence graphs \cite{Grasso_2004}, sequences to cyclic graphs \cite{Navarro_2000}, and even cyclic sequence graphs to cyclic sequence graphs \cite{Myers_1989, Amir_1997}.
It is notable that many of these findings have been independently rediscovered or refined by contemporary researchers \cite{Antipov_2015, Rautiainen_2017, Jain_2019a,kural2014methods}.
Some earlier algorithms require restricted scoring functions to achieve efficiency \cite{Rautiainen_2017}, but recent contributions have used less restricted functions that produce more biologically meaningful alignments in some contexts \cite{Jain_2019a}.

Graph alignment algorithms have also become faster.
\textsc{POA} had equivalent asymptotic run time to linear alignment but required acyclic graphs \cite{Lee_2002}. 
Later optimizations simply ran slower on general graphs \cite{Kavya_2019}.
Algorithms are now known with equivalent run time even on general graphs \cite{Jain_2019a}.
In addition, researchers have developed modified algorithms that run quickly in the practical context of real-world computer architectures \cite{Suzuki_2018, Rautiainen_2019, Jain_2019b}.


\subsection{Genome graph mapping}
\label{sec:graphmapping}

Efficient methods to map reads to large pangenome graphs have been developed in recent years.
Many draw on recent research in alignment (\S \ref{sec:graphalignment}), and advances in pangenome indexing (\S \ref{sec:indexing}).

Although these mapping tools all target sequence graphs, there are significant differences in the types of graphs that they handle.
Several tools apply only to acyclic variation graphs formed by adding variants to a linear reference.
Examples include \textsc{GenomeMapper} \cite{Schneeberger_2009}, Seven Bridges' \textsc{Graph Genome Aligner} \cite{Rakocevic_2019}, \textsc{HISAT2} \cite{Kim_2019}, \textsc{V-MAP} \cite{Vaddadi_2019}.
In contrast, \textsc{VG} \cite{Garrison_2019} and \textsc{GraphAligner} \cite{Rautiainen_2019b} appear to be the only tools with open ambitions of mapping to arbitrary variation graphs, including complex local and global topologies.
\textsc{GraphAligner} can also align to generic overlap graphs and DBGs, a feature which it uses to drive the error correction of long reads using DBGs \cite{Heydari_2018, Fu2019ErrorCorrectionSurvey}.

The majority of these tools emphasize mapping short-read \emph{next-generation sequencing (NGS)} data. 
To our knowledge, \textsc{GraphAligner} and \textsc{V-MAP} are the only graph mapping tools designed long read sequencing data \cite{Rautiainen_2019b, Vaddadi_2019}.
While \textsc{V-MAP} also supports NGS reads, \textsc{GraphAligner}'s seeding strategy limits it to long reads.
\textsc{VG} also supports long-read alignment \cite{Garrison_2018}, but this is based on a hierarchical approach that applies the alignment algorithm for short reads to chunks of long reads \cite{Garrison_2019}.
The approach is accurate but nearly an order of magnitude slower than \textsc{GraphAligner} \cite{Rautiainen_2019b}.

For indexing (\S \ref{sec:indexing}), most graph mapping tools have opted for some variation of a $k$-mer table. 
\textsc{GraphAligner}, \textsc{GenomeMapper}, Seven Bridges' mapper, and \textsc{V-MAP} all use this strategy \cite{Rautiainen_2019b, Schneeberger_2009, Rakocevic_2019, Vaddadi_2019}. 
The remaining mappers use succinct text indexes.
\textsc{VG} uses the GCSA2 \cite{Siren_2017} and a longest-common-prefix array, which enable very specific queries at the expense of high memory utilization \cite{Garrison_2019}.
\textsc{HISAT2} uses a modified GCSA \cite{Siren_2014} that also encodes the graph structure itself.
This helps give \textsc{HISAT2} an impressively low memory footprint but a somewhat more limited set of queries \cite{Kim_2019}.
\textsc{GraphAligner} also has the option of seeding with a full text index of the node sequences of the graph, but this is not enabled by default \cite{Rautiainen_2019b}.

Most graph mappers employ graph-based alignment algorithms. 
The exceptions are \textsc{GenomeMapper}, which aligns to all paths out from a seed, and \textsc{HISAT2}.
The \textsc{HISAT2} alignment algorithm relies on a complex set of heuristics that depend heavily on its exact match index.
This makes it exceptionally fast, although it also can hurt alignment quality around indels. 
\textsc{VG} and \textsc{V-MAP} both employ some version of partial order alignment \cite{Garrison_2019, Vaddadi_2019}.
Seven Bridges first searches for a near-exact match using an exponential depth-first search, applying partial order alignment if this fails \cite{Rakocevic_2019}.

Due to the recent development of these methods, there have been few independent comparative studies of their performance and accuracy.
In general, \textsc{VG} compares favorably to other tools in terms of accuracy on NGS data \cite{Rand_2017}. 
However, it requires more memory, has slower indexing, and often maps more slowly than the alternatives \cite{Kim_2019, Vaddadi_2019}. 
\textsc{V-MAP}'s fast clustering heuristics allow it to align long reads faster than \textsc{VG}, but it was not compared to \textsc{GraphAligner} \cite{Vaddadi_2019}.
\textsc{GraphAligner} is the only mapper to incorporate the most recent research into graph alignment algorithms.
It uses a banded alignment algorithm to achieve impressive speed aligning long reads to genome graphs \cite{Rautiainen_2019b}.

RNA splicing can be represented directly in a genome graph model \cite{Lee_2002}.
It follows that graph mappers can be applied to RNA sequencing data.
\textsc{HISAT2} can map RNA-seq data in addition to genomic DNA \cite{Kim_2019}.
It is based on the RNA mapper \textsc{HISAT} \cite{Kim_2015}, and it retains the capacity for spliced alignment.
The ability to create spliced variation graphs has also been added to VG \cite{Garrison_2018}. 
In these variation graphs, known splice junctions are added as edges, similar to the addition of a deletion event.
\textsc{VG} supports any type of variation, but its splicing-awareness is limited to splice junctions represented in the graph.
Thus, reads that span a novel splice junction will only map partially.
We further discuss applications of graph mapping to functional genomics in \S \ref{sec:functionalgenomics}.


