\section{Relating new information to the pangenome}

\subsection{Visualization}
% Adam

Once one has constructed a graph pangenomic model, one might want to look at it.
A number of tools have been developed for this purpose.
\todo[inline]{JME: i think this section would benefit from a more inspirational lead-in.
we want people to think that graph visualization is important, not just a thing that exists}

Bandage (\textbf{B}ioinformatics \textbf{A}pplication for \textbf{N}avigating \textit{\textbf{D}e novo} \textbf{A}ssembly \textbf{G}raphs \textbf{E}asily) \citep{Wick_2015} is one of the most popular \citep{Mikheenko_2019}.\todo{JME: IMO, spelling out these acronyms breaks up the flow and is not very important.
bioinformatics acryonyms are shit, we don't need to elevate them in the text.}
Originally designed for working with bacterial assembly and meta-assembly graphs \citep{Wick_2015}, it supports a wide range of formats and graph  paradigms.
It can be effectively used for interactively visualizing and exploring subregions of human-scale pangenome graphs \citep{Garrison_2019}, but its shortcomings become apparent at larger scales or with higher degrees of connectivity between graph regions \citep{Mikheenko_2019}.
The tool does have support for restricting the portion of the graph displayed to a particular ``scope'', but the tool is fundamentally built around laying out the graph under study in two dimensions, with all nodes represented, and panning and zooming around it.
Moreover, while the tool includes the ability to search for sequences in the graph, it does not have any ability to structure graph display using known linearization information.
Additionally, Bandage is implemented as a cross-platform native C++/Qt application \citep{Wick_2015}, which allows the tool to be self-contained but precludes the use of cloud resources for dealing with larger graphs; all graph data must be stored and processed on the user's machine at visualization time.
\todo[inline]{JME: i'm worried some of these implementation details might be too in-the-weeds.
they have important implications, but i think this discussion would be more accessible to a general audience if we refocus it on the effects that these implementation decisions have on the user }

GfaViz is another C++/Qt application for visualizing genome graphs which claims full support for newer GFA2 features, such as the gaps which allow GFA2 to represent scaffold graphs in addition to assembly graphs; these features are not available in Bandage, which supports only GFA1 \citep{Gonnella_2018}.
However, unlike Bandage, the GfaViz project has not released any binary builds or user interface screenshots of their application, so Bandage remains the more accessible tool.
\todo{JME: we should probably introduce the GFA format before mentioning it.
i wonder where we should do that...}

Departing from the desktop application model, two recent graph visualization tools, \textbf{S}caffold \textbf{G}raph \textbf{T}ool\textbf{K}it (SGTK) and \textbf{A}ssembly \textbf{G}raph \textbf{B}rowser (AGB), adopt a build-a-web-page model, in which a web-based visualization is prepared that does not rely on further server support, lowering end-user system requirements \citep{Kunyavskaya_2018,Mikheenko_2019}.
SGTK is designed for visualizing scaffold graphs, which can have negative-overlap gaps between sequenced segments, and is based around the in-browser Cytoscape.js graph layout and rendering library \citep{Kunyavskaya_2018}.
It includes a potentially highly interpretable, reference-sequence-structured ``browser'' layout, but its use of Cytoscape.js appears to push it towards a generic circular node representation, rather than the noodles of Bandage \citep{Kunyavskaya_2018}.
\todo{JME: i love `noodles', but i don't know if it's a science word}
Moreover, its authors do not demonstrate its usability on large graphs, with the largest graph evaluated having only 923 nodes and 60,679 edges \citep{Kunyavskaya_2018}.

AGB has a similar overall design to SGTK, but makes different implementation choices.
It is relatively tightly tied to the assembly graph use case, requiring each (sequence-bearing) edge to be classifiable as ``unique'' or ``repetitive'' based on assembler annotation or sequencing coverage \citep{Mikheenko_2019}.
Where SGTK used Cytoscape.js, AGB relies on the venerable \texttt{graphviz} tool itself, compiled for execution in the browser \citep{Mikheenko_2019, Ellson_2001}.
This lets it lay out graphs in a more flow-guided way, using graphviz's rank-based algorithms, as compared to Cytoscape.js's more force-directed-looking layouts \citep{Mikheenko_2019, Kunyavskaya_2018}.
AGB is built around the idea of splitting up an assembly graph and visualizing portions of it, and features many ways to do this, including linear-reference-based and minimum-edge-cut-based approaches \citep{Mikheenko_2019}.
On the backend, the tool is built around potentially megabyte-scale JSON files\footnote{\url{https://github.com/almiheenko/almiheenko.github.io/blob/8f4b2f8c7c498f04fa32f53f69b4bc59888a14f0/AGB/Flye_Human/data/repeat_graph.json}}, with no apparent provision for region-specific download, but the tool is still demonstrated to be scalable enough to handle human and other eukaryotic assembly graphs \citep{Mikheenko_2019}.

There is a difference in scale when moving from an assembly or scaffold graph to a comprehensive pangenome graph for even a species with as little diversity as humans.
While tools like SGTK and AGB have been demonstrated on graphs with tens of thousands of entities, the 1000 Genomes Project dataset contains 88 million known human variants \citep{1000_2015}, which gives a density over the 3 billion base human genome of about 34 bases per variant, and a comprehensive pangenome graph with tens to hundreds of millions of elements---a much larger graph than those that SGTK and AGB have been shown to work with.
Moreover, to achieve their single-megabyte-scale visualization file sizes for hundred-megabase- to gigabase-scale assemblies, these tools necessarily elide sequence information.

To deal with comprehensive pangenome graphs with sequence data, one design approach is to keep the browser-based client but to put more intelligence into the server.
This is the avenue taken by the Sequence Tube Map, which renders regions of pangenome graphs using a visual language inspired by transit system maps \citep{Beyer_2019}.
This tool operates at a much more magnified zoom level than tools designed to work with assembly graphs, and is useful for visualizing base-scale variation and short read mapping locations in human-chromosome-scale graphs \citep{Beyer_2019}.
However, its imposition of a local linear ordering and its limited graph simplification tools make it difficult to use on larger regions \citep{Beyer_2019}.
Additionally, its architectural decision to load the graph from disk for every request makes latency prohibitively high when working with combined graphs above the scale of a single chromosome, so the tool is not suitable for graphs which cannot be broken up into chromosome-scale connected components.

Other approaches to visualizing comprehensive pangenome graphs rely on restricting the problem in order to improve efficiency.
For example, by forcing the sequence-bearing nodes of the graph into a one-dimensional layout, and producing vector rather than raster output, the \texttt{vg viz} tool promises to render graphs of arbitrary size and complexity in linear time \citep{Garrison_2019}. 
\todo{JME: i don't know if our audience will understand the distinction between raster and vector}
The intended use case seems to be to use a single global linearization, and to interactively render regions of it for display, but it is also possible to non-interactively render an entire graph to a very large image \citep{Garrison_2019}. 
The tool is designed around a base-level visual representation of the graph, and optimized for comparing long embedded paths in the graph, on the basis of which nodes they do or do not visit \citep{Garrison_2019}. 
However, as with the Sequence Tube Map, it can be difficult to understand high-level or nonlinear structures in such a linearized layout \citep{Garrison_2019}. 
In contrast to the Sequence Tube Map no provision is made for visualizing short read alignments \citep{Garrison_2019}. \todo{EG: actually, this isn't exactly true. We do support coverage of short read alignments over the graph (it can read in the .pack format).}

Figure~\ref{fig:visualization} provides a visual overview of the visualization methods surveyed here.
It remains an open problem to interactively visualize a comprehensive human pangenome graph, potentially including interchromosomal connections, at a variety of zoom levels, with modestly provisioned end-user hardware.

\begin{figure}[h]
    \includegraphics[width=0.9\textwidth]{figures/visualization.pdf}
    \caption{\label{fig:visualization} An overview of several approaches to visualizing assembly, scaffold, and comprehensive pangenome graphs.
\textbf{A:} Bandage, adapted from \cite{Wick_2015} supplementary section~6.
\textbf{B:} GfaViz, adapted from \cite{Gonnella_2018} supplementary figure~S4.
\textbf{C:} SGTK, adapted from \cite{Kunyavskaya_2018} figure~1.
\textbf{D:} AGB, adapted from \cite{Mikheenko_2019} supplementary figure~S3.
\textbf{E:} Sequence Tube Map, adapted from \cite{Beyer_2019} figure~2.
\textbf{F:} \texttt{vg viz}, adapted from \cite{Garrison_2019} figure~2.20.}
\end{figure}

% Go back and talk about ABySS-Explorer (2009) and its "polar" graphs and wiggly sequence edges?
% Bandage: cross-platform native application
    % Versatile and popular
    % Client-only
    % Could look up a reference but not view vs it
% GfaViz: C++/Qt tool with full GFA1/2 support
    % Has a GUI but no screenshots or binaries
    % Can show cool stuff like reads connected to their assembly contigs
% SGTK
    % build-view model
    % Cytoscape.js or genome browser linear-structured
    % designed for scaffold graphs (more processed?)
    % Not proven on large graphs; only shown going up to 100s of nodes
% AGB: auto-subgraphs (to 100 nodes) and simplifies assembly graphs
    % js GraphViz based
    % Still uses a build-web-page model
    % Kind of tied to assemblies (notion of repetitive vs non-repetitive edges)
    % Can view vs a reference
    % Scales to C elegans at least
        % O(300 * 100 = 30000) nodes
    % Appears to be structured around megabyte-scale GraphViz graphs https://github.com/almiheenko/almiheenko.github.io/tree/master/AGB
        % No LOD-ing on the backend for efficient download, but not a problem at the scale of 10s of thousands.
% Tube Map
    % Client-server model: requires a server, but can use server resources to crunch the graph
    % Designed to impose a left to right local ordering that orients the edges in a hopefully sensible way
    % Scales well to millions of nodes, demonstrated on partial human pangenome graph references


%\citep{Wick_2015} : Bandage
%\citep{Gonnella_2018} : GfaViz
%\citep{Kunyavskaya_2018} : SGTK
%\citep{Mikheenko_2019} : AGB
%\citep{Beyer_2019} : TubeMap
%\citep{Garrison_2019} : Thesis

\subsection{Finding structures in pangenome graphs}
% Jordan

\todo[inline]{JME: We need to decide whether to do this section}

\subsection{Graph alignment algorithms}

Often, genomic sequence data can only be interpreted in the context of other sequences. 
For this reason, sequence comparison is at the core of many genomic analyses, and sequence alignment is the essential method for doing so. 
However, classic alignment algorithms like Smith-Waterman\cite{Smith_1981} do not directly apply to sequence graphs. 
Accordingly, the increasing prominence of variation graph methods has been fueled by fundamental algorithmic research in graph alignment, and it has also spurred further research.
\todo{JME: this lead-in is redundant with the introduction, but it could be rewritten to tie into that content.
i'll do that later.}

The trend in the graph alignment research is toward greater generality and faster run time. 
The generality comes in two main forms. 
First, algorithms apply to increasingly general classes of graphs. 
The foundational genomic sequence graph alignment algorithms applied only to graphs without cycles\cite{Lee_2002, Grasso_2004}.
\todo{EG: this is not accurate. Myers showed cyclic to cyclic graphs in 1990. This is also described in \emph{Biological sequence analysis}.}
More recent research has discovered algorithms that align to graphs with any shape\cite{Antipov_2015,Rautiainen_2017,Jain_2019a}. 
Second, graph alignment researchers have developed algorithms that use increasingly general scoring functions. 
Some earlier algorithms require restricted scoring functions to achieve efficiency\cite{Rautiainen_2017}, but recent contributions have used the less restricted scoring functions that are required to produce biologically meaningful alignments in many contexts\cite{Jain_2019a}.

Graph alignment research also has improved the algorithms' run time. 
The first algorithms only required restricted graphs in order to run at comparable speed to non-graph sequence alignment algorithms, or else simply ran slower on general graphs\cite{Lee_2002, Kavya_2019}. 
It has now been shown that graph alignment can run at the same speed as non-graph alignment (in the computer scientist's sense of ``Big-O'' asymptotics).\todo{EG: Let's just use the actual notation here.}
This is believed to be essentially optimal\cite{Jain_2019a,Equi_2019}. 
Many of the advances in this space came from rediscovering analogs to the graph alignment problem in the related areas of regular expressions and hypertext\cite{Myers_1989,Amir_1997}. 
In addition to these theoretical results, researchers have also developed modified algorithms that run quickly the practical context of real-world computer architectures\cite{Suzuki_2018,Rautiainen_2019,Jain_2019b}.

From a practical standpoint, the primary benefit of this research in alignment algorithms has been in aiding the design of mapping tools. 
Graph alignment algorithms are a central component of the graph mapping tools described below.

\subsection{Variation graph mappers}

\todo[inline]{JME: I think this section is missing a brief introduction to the seed and extend paradigm.
that's probably where the discussion should start.}
There are several mapping algorithms for variation graphs which extend concepts from linear mapping algorithms into a graph space.
Variation graph mapping algorithms generally fall into one of two categories: kmer-based and Burrows-Wheeler Transform-based.
\todo{JME: we need to harmonize the formatting of `k-mer' across the sections}
Kmer-based approaches for mapping to variation graph use an index that maps kmers in the graph to their location.
These algorithms generally follow the same structure.
Kmers from the read are looked up in the index and the hits, matches between the read and the graph, are turned into seed alignments which may be clustered by their position on the graph to get a subgraph to align against.
\todo{JME: why only discussing k-mers here when you also brought up BWT?}
Once the read has been mapped to one or more subgraph or regions on the graph, an alignment algorithm is used to get the final alignment to a sequence in the graph.

kmer-based graph mapping algorithms diverge in how kmers are indexed, how seeds are defined, how a subgraph is chosen to align to, and which alignment algorithm is used.
\todo{Indexing is maybe talked about in a different section?}
Once hits have been found in the kmer index, overlapping and adjacent hits can be merged and/or extended to produce seed alignments.
Different types of seeds (NIMSs, MEMs, SMEMs, MUMs, minimizers) are each defined by a different set of properties such as the number of gaps and mismatches allowed and how often they occur in the reference.
\todo{JME: these are technical abbreviations that we can't assume the readers will know.
really i think the commonality is that they're exact matches.}
Seeds can be clustered by their proximity on the graph and clusters with that are less likely to produce a good alignment can be filtered out at this step.
Clusters can define a subgraph that the read gets aligned to.
Clustering and the repeated distance calculations it involves can be a prohibitively slow process and it is omitted in some algorithms.
Alignment, either to the graph itself or to a subgraph, can be done using the algorithms described above.
\todo[inline]{JME: despite how it's introduced a lot of this content applies to non-k-mer seeds as well.
some of this content could be reworked into a good introductory paragraph for this section}

HISAT2 is a BWT-based mapping algorithm that uses an adaptation of the Ferragina-Manzini (FM) index, a hierarchical graph FM index (HGFM).
\todo{JME: it actually uses a GCSA, not a standard FM index}
An FM index is a compressed representation of the graph that is optimized for substring searching.
The HGFM is comprised of two FM indexes: a global index of the entire genome and local indexes of smaller portions of the genome and their variants.
Repeat sequences are combined into a separate index, keeping only one copy of each repeat so sequences in repetitive regions are only aligned once.
Alignment is done by searching the FM indexes, starting with the global index.


Each of these graph alignment tools has demonstrated an increase in sensitivity over mapping to a standard linear reference.
GenomeMapper, vg map, Seven Bridges' mapping algorithm, and HISAT2 are primarily short read mappers.
\todo{JME: i think 7bg's tool is called Graph Genome Aligner}
GraphAligner is a long read mapper and V-MAP is used for both.
V-MAP doesn't align reads itself but finds a small subgraph that a read can be aligned to using existing mappers. 
HISAT2 and vg can align both RNA and DNA.


\subsection{De Bruijn graph mappers}
% Adam

% Style Note: apparently de Bruijn Graph should be lower case unless otherwise required to be capitalized.

In addition to mappers designed to map to general graphs, a selection of mappers designed to map specifically to de Bruijn graphs, or to compacted de Bruijn graphs\todo{Did we define this already?}, have also been developed.

One of the first of these, BlastGraph from 2012, uses an edit distance alignment metric for scoring, and finds all matches in the graph within that edit distance \citep{Holley_2012}.
However, it seems better suited to search than to mapping to support resequencing; but it is only shown running on up to 10,000~reads at a time \citep{Holley_2012}.
The actual alignment algorithm uses a hash-table-based seed index combined with a recursive-depth-first-search-based extension step, and is shown applied to de Bruijn graphs and ordinary sequence graphs \citep{Holley_2012}.
\todo{Can we move this to the previous section since it doesn't really do anything de-Bruijn-specific?}

\citeauthor{Holley_2012} do not discuss the algorithmics of their alignment approach, beyond demonstrating that it is sufficiently fast in practice \citep{Holley_2012}.
In \cite{Limasset_2016}, \citeauthor{Limasset_2016} present a proof that the ``de Bruijn Graph Read Mapping Problem'' (when looking for acyclic mappings) is NP-complete, by a reduction of the Hamiltonian Path Problem, through a Traveling Salesman Problem variant, and then through a gapless, acyclic read-to-sequence-graph mapping problem \citep{Limasset_2016}.
This realization of NP-completeness goes on to structure future work in the field.
See also, however, the Dijkstra's-algorithm-based polynomial-time solution given in \cite{Antipov_2015} to the problem of finding the optimal path between a particular anchoring source and sink.

\citeauthor{Limasset_2016} present a greedy tool, BGREAT, which avoids the theoretically exponential recursive-depth-first-search step of BlastGraph, and which leverages the overlaps between nodes in the compacted de Bruijn graph as seeds \citep{Limasset_2016}.
This seed selection mechanism inherently limits the number of seed hits which need to be indexed, and provides (modulo errors) some intuitive guarantees about the distribution of seeds in a true mapping \citep{Limasset_2016}.
However, the authors compare BGREAT only against Bowtie2, and it is unclear how to interpret their comparison results, as Bowtie2 is, with BGREAT, part of their larget GGMAP, and moreover does not align to graphs \citep{Limasset_2016}.

While BGREAT is motivated in terms of mapping to assembly graphs \citep{Limasset_2016}, the deBGA tool, also published in 2016, is aimed at mapping to ``Reference de Bruijn Graphs'' describing variation within or between species \citep{Liu_2016}.
The mapping algorithm uses the $k$-mers of the de Bruijn graph as seeds, and unlike BGREAT and BlastGraph the core alignment operation is sequence-to-sequence \citep{Liu_2016}.
Also unlike BGREAT and BlastGraph, however, deBGA provides a paired-end mapper \citep{Liu_2016}.
Speed and accuracy compare favorably with linear-reference-based tools, but again no comparison against other graph-based tools is made \citep{Liu_2016}.
\todo{The BrownieAligner paper accuses BGREAT of not supporting indels; BGREAT says it allows an ``edit or Hamming distance'', which is ambiguous between supporting both or characterizing edit distance as just meaning Hamming distance.}

A more recent read to graph mapping tool, 2018's BrownieAligner, is compared against both BGREAT and deBGA; although it can be slower, it is more accurate, because it is able to compute genuinely optimal alignments in reasonable time \citep{Heydari_2018}.
It accomplishes this by using a branch-and-bound approach to limit its depth-first search out from each de Bruijn $k$-mer or fallback MEM seed \citep{Heydari_2018}.
Additionally, BrownieAligner allows a ``higher-order Markov model'', specifying the probability of each de Bruijn graph node as a function of the previous $n$-node path taken through the graph, to constrain the possible paths through the graph against which reads are aligned \citep{Heydari_2018}.
This is similar to the idea behind using haplotypes to inform mapping in sequence graphs \citep{Siren_2019}, except that in BrownieAligner the Markov model describing acceptable paths is learned from a first pass of read alignment, after which the sufficiently-covered paths in the graph become the paths in the Markov model used to restrict a second alignment pass \citep{Heydari_2018}.
In this way, some variant- or sample-level information can be extracted from the read set as a whole and fed back into the individual read mappings, improving their accuracy \citep{Heydari_2018}.



% Useful to map to dBGs

% BlastGraph
    % Works on dBGs and other SGs
    % Find paths within edit distance of query
    % Hash table index of all seeds of a certain length
        % Stores only start positions
    % Compute edit distances and alignments out in both directions from each seed
        % Deduplicate when multiple seeds suggest the same alignment
    % Wrote a Java Cytoscape plugin and a C implementation

% BGREAT
    % Works on dBGs and actually maps
    % Justified in terms of mapping to assemblies
    % Mapping to dBGs is NP-complete even without gaps
        % How does this square with the edit distance algorithm's algorithmics?
            % Finding all the seed starts in the dBG for BlastGraph could be exponential for seeds longer than nodes
            % Recursive DFS tail edit distance alignment could also be exponential
            % But dBG overlap structure limits the density of braiding
        % The proof is by reducing Hmailtonian Path (path that visits each node once) -> restricted Traveling Salesman -> general graph mapping -> mapping to dBG
    % Actual tools: GGMAP (BGREAT + Bowtie2 for unitigs), and BGREAT for branching paths
        % BGREAT uses the dBG overlap regons themselves as seeds
        % They think they can get away with only mapping with Bowtie2 if they don't find a good mapping with BGREAT
            % They might be right; the dBG structure kind of precludes any competitive mappings in the middle of unitigs when there's a good mapping that covers an overlap
                % Because if there was another similar place in the genome it would have an overlap with the stuff it is similar to
        % Uses a greedy threading approach starting from the outer overlap sequences that occur in the read
            % Drop the whole read if it can't find a mapping for each end from the outermost overlap seeds.
        % Get linear time mapping which is nice
        % Not quite clear to me how they did the comparisons: are they just giving the time that their bowtie2 and blastgraph steps took on the reads that were fed to them?
            % Maybe they didn't find other comparable tools to compare to.

% deBGA
    % Justified in terms of mapping to variation and MSAs
    % "Putative Read Position" (PRP) (basically read start position/seed offset) 
    % Uses MEM-based seeding
    % Alignment is either exact match extension or aligning to "clustered" sequences, no string-to-graph
    % Paired end support (which BGREAT and BlastGraph don't mention)
    % Very fast, even on human
    % Not clear how reads are surjected to linear space for comparison
        % "Multiple equally best alignments"
            % deBGA and BWA with 1000 max mappings
        % If not surjected the problem is easier (no need to try to resolve repeats)

% hybridSPAdes
    % This is a short-and-long-read assembler, not a mapper
    % Maps long reads into a short-read dBG-derived graph
        % Just looks for 8 or more matching 13-mers to an "edge" for "support"
        % Does exponential string to graph alignment for the hard parts
        % Propose a polynomial algorithm for aligning between anchored ends for minimum edit distance

% BrownieAligner
    % Branch and bound approach to seed and extend
    % "Higher order markov model" to mark paths in the graph that exist, like our GBWT
        % Two pass approach where they train the markov model based on one batch of alignments and then realign
        % I think this is what lets them learn novel paths not in the linear reference but still catch errors
        % "Markov models offers a significant improvement for the alignment of these harder to align reads."
    % Lets seeds be smaller then dBG kmer size
    % "for each dataset the best accuracy for BrownieAligner is always higher than the best accuracy for other tools"

% Really comparing and contrasting the empirical speed, accuracy, etc claims of these tools would be hard and really requires independent experiments
    % Although BrownieAligner compares against BGREAT and deBGA
    

% BlastGraph (2012) \citep{Holley_2012}
% BGREAT (2016) \citep{Limasset_2016}
% deBGA (2016) \citep{Liu_2016}
    % deBGA-VARA is a variation-aware extension but not published open-access
% hybridSPAdes (2016) \citep{Antipov_2015}
% BrownieAligner (2018) \citep{Heydari_2018}

% See also: https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-1103-9


\subsection{Non-graph population mapping tools}
% Erik

